---
title: "Pseudo-bulk peaks K27 HBCx95 UNT m43"
author: "CÃ©line Vallot"
date: "05/06/2019"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
library(scran)
library(corrplot)
library(devtools)
#library(BiocInstaller)
library(pheatmap)
library(limma)
library(scater)
library(statmod)
#library(pcaMethods)
library(Matrix)
library(dplyr)
library(geco.utils)
library(geco.visu)
#library(ConsensusClusterPlus)
library(geco.unsupervised)
library(geco.ChIPseq)
#library(seriation)
#library(heatmap.plus)
#library(scatterplot3d)
library(Rtsne)
library(RColorBrewer)
library(colorRamps)
library(dplyr)
library(tidyr)
library(viridis)
library(wesanderson)
library(knitr)
library(ChromSCape)
options(stringsAsFactors = F)
```

```{r, echo=FALSE}
#Directories
dataDir <- "HBCx95_K27_scBulk/07csv/"
annotDir <- "HBCx95_K27_scBulk/peaks/"
resdir1 <- "HBCx95_K27_scBulk/07csv/"

#Count matrices
resultZ <- read.csv(file.path(dataDir,"results_Zerone_K27_macs2.csv"))
rownames(resultZ) <- paste(resultZ$Chromosome,resultZ$Begin,resultZ$End,sep="_")

#Variables
annotCol <- c("Treatment","Exp")
```


```{r, echo=FALSE}
#Annotation of peaks with Gencode info, K4me3 peaks and enhancers of breast cancer
annotZerone <- read.table(file.path(annotDir,"macs2_peak_call_K27.bed"),sep="\t")
colnames(annotZerone) <- c("chr","start","end")
annotZerone$ID = paste0(annotZerone$chr,"_",annotZerone$start,"_",annotZerone$end)
rownames(annotZerone) <- annotZerone$ID
annotZerone$lengthPeak <- annotZerone$end - annotZerone$start

annotZerone_tss_all_i <- read.table(file.path(annotDir,"HBCx95_UNT_K27.TSS.bed"),sep="\t")[,-c(4:7,9)]
colnames(annotZerone_tss_all_i) <- c("chr","start","end","Gene","distance")
annotZerone_tss_all_i <- annotZerone_tss_all_i[!duplicated(annotZerone_tss_all_i),]
annotZerone_tss_all_i$ID = paste0(annotZerone_tss_all_i$chr,"_",annotZerone_tss_all_i$start,"_",annotZerone_tss_all_i$end)
annotZerone_tss_all <- aggregate(Gene ~ ID, data = annotZerone_tss_all_i, paste, collapse = ",")
annotZerone_tss_all$distance <- aggregate(distance ~ ID, data = annotZerone_tss_all_i,min )[,2]
annotZerone$transcript_all <- ""
annotZerone$transcript_all <- annotZerone_tss_all$Gene[match(annotZerone$ID,annotZerone_tss_all$ID)]
annotZerone$distance_all <- ""
annotZerone$distance_all <- annotZerone_tss_all$distance[match(annotZerone$ID,annotZerone_tss_all$ID)]
annotZerone$transcript_close = ""
annotZerone$transcript_close[which(annotZerone$distance_all<1000)] = annotZerone$transcript_all[which(annotZerone$distance_all<1000)]

annotZerone_K4 <- read.table(file.path(annotDir,"HBCx95_UNT_K27.K4.enlarged5000bp.bed"),sep="\t")
colnames(annotZerone_K4) <- c("chr","start","end")
annotZerone_K4$ID_K4 <- paste(annotZerone_K4$chr,annotZerone_K4$start,annotZerone_K4$end,sep="_")
annotZerone$K4_ID <- annotZerone_K4$ID_K4[match(annotZerone$ID,annotZerone_K4$ID)]
annotZerone$K4 <- !is.na(annotZerone$K4_ID)

table(annotZerone$K4)
annotZerone$Gene = annotZerone$transcript_close
annotZerone$bivalent <- (!is.na(annotZerone$Gene) & annotZerone$K4)
annotZerone <- annotZerone[rownames(resultZ),]
```

```{r, echo=FALSE}
#Preparing raw count and normalized count matrices
rownames(resultZ) <- paste(resultZ$Chromosome,resultZ$Begin,resultZ$End,sep="_")
RZ <- (resultZ[,-c(1:3,grep(pattern = "input",colnames(resultZ))),drop=F]) #RZ matrix is the raw count matrix

#create RPKM matrix by normalizing per library size and peak length
RZ_RPKM <- (10^9*t(t((RZ+1)/annotZerone$lengthPeak)/colSums(RZ)))
```


```{r, echo=FALSE}
#Selecting samples for unsupervised analysis

#RZ <- (RZ[,-grep(pattern = "sc",colnames(RZ))]) #remove single cell datasets from analysis
RZ_sel <- RZ
colnames(RZ_sel)
ExpName <- "HBCx95_UNT_K27"

dim(RZ_sel)
annotZerone_sel <- annotZerone; dim(annotZerone_sel)

exp.sd <- apply(RZ_sel, 1, sd)
sel2 <- order(exp.sd, decreasing = TRUE)[1:1000]
RZ_sel2 <- RZ_sel[sel2,,drop=F]; dim(RZ_sel2)
annotZerone_sel2 <- annotZerone_sel[sel2,,drop=F]

mat <- log2(RZ_RPKM[rownames(RZ_sel2),colnames(RZ_sel2),drop=F])
dim(mat)
hist(mat,breaks=100,main="Distribution of log2 pseudo IP")

```

```{r, echo=FALSE}
library(scTools)
  resSUBdir <- file.path(resdir1, "../Plots") ; if(!file.exists(resSUBdir)){dir.create(resSUBdir)}
  RDatadir <- file.path(resSUBdir,"RData") ; if(!file.exists(RDatadir)){dir.create(RDatadir)}
  annotSample <- data.frame(
    "Name" = colnames(mat),
    "Treatment"= "UNT")
  
  annot_sel = annotSample
  

  anocol <- geco.annotToCol2(annotS=annot_sel[,c("Name","Treatment"),drop=F],
                             annotT=annot_sel,plotLegend=T,
                             plotLegendFile=NULL, categCol=NULL)
  
```

5. Comparison K27me3 and K4me3 signal
```{r, echo=FALSE}
load("../../Raw_analysis/scRNAseq/Results_PDX/Supervised/RData/Supervised_res_object_edgeR.Rdata")

res= my.res

#Diff
diff_res = res[which(res$log2FC.persister_6_vs_UNT > 2 & res$qval.persister_6_vs_UNT < 0.1 ), ]
rownames(diff_res) = diff_res$Symbol

annotZerone. = annotZerone %>% tidyr::separate_rows(Gene,sep=",")
annotZerone. = annotZerone.[which(annotZerone.$Gene %in% diff_res$Symbol),]
annotZerone. = annotZerone.[which(annotZerone.$distance_all <1000),]

contingency_table_diff = data.frame(OverlappingPersister = table(annotZerone.$K4)[2:1],
                                Total= table(annotZerone$K4[which( (annotZerone$distance_all < 1000) & !is.na(annotZerone$Gene)) ])[2:1],
                                row.names = c("Bivalent","Non bivalent"))[,c(2,4)]
contingency_table_diff
load("../../Raw_analysis/scRNAseq/common_over_genes_pers_vs_unt.RData")
annotZerone. %>% dplyr::filter(Gene %in% common_overexpressed_genes)

(fdiff = fisher.test(contingency_table_diff))
 
vec = c(as.numeric(contingency_table_diff[1,]/(contingency_table_diff[1,]+contingency_table_diff[2,])))
df = data.frame(percentage_bivalent = round(100*vec[2:1]),type = factor(c("Total","PersisterGenes"),levels = c("Total","PersisterGenes")))

library(ggpubr)
stat.test = data.frame(.y. = "percentage_bivalent",
                       group1 = "Total" ,
                       group2 = "PersisterGenes",
                       p = fdiff$p.value,
                       p.adj = fdiff$p.value,
                       p.format = fdiff$p.value,
                       p.signif = "****",
                       method="Fisher Test")


  png(file.path(resSUBdir,"Bivalent_enrichment_in_persisterGenes.png"), res=300, width = 800, height = 1500)
  df %>% ggplot(aes(x=type,y=percentage_bivalent)) + geom_bar(aes(fill=type),stat ="identity", width = 0.5) + stat_pvalue_manual(data=stat.test,label = "p", y.position = 35) + ggtitle(paste0("Bivalent peaks enriched in H3K27me3 differential peaks (5FU vs DMSO)")) +
      theme_classic() + scale_fill_manual(values = c("#000000ff","#ff0000ff"))  +
      theme(axis.text.x = element_text(angle = 90, vjust =1), legend.position = "None") +
      ylab("Proportion of bivalent peaks (%)") + xlab("") + ylim(c(0,50))
  
  dev.off()

```



