---
title: "MM468 K4 transcripts 10k bulk"
author: "CÃ©line Vallot"
date: "05/06/2019"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
library(scran)
library(corrplot)
library(devtools)
library(pheatmap)
library(limma)
library(scater)
library(statmod)
library(Matrix)
library(dplyr)
library(geco.utils)
library(geco.visu)
library(geco.unsupervised)
library(geco.ChIPseq)
library(Rtsne)
library(RColorBrewer)
library(colorRamps)
library(dplyr)
library(tidyr)
library(viridis)
library(wesanderson)
library(knitr)
library(ChromSCape)
options(stringsAsFactors = F)
```

```{r, echo=FALSE}

#Directories
dataDir <- "K4_transcripts_10k/07csv/"
annotDir <- "peaks/"
resdir1 <- "K4_transcripts_10k/results/"

#Count matrices
#result50k <- read.csv(file.path(dataDir,"results_hg38.50k.csv"))
resultZ <- read.csv(file.path(dataDir,"results_MM468_all_transcripts10k_v2.csv"))
resultZ_CI <- read.csv(file.path(dataDir,"MM468_chromInd_by_index_K4_transcripts_10k_v2.csv"))
resultZ = cbind(resultZ,resultZ_CI[,-c(1,2,3)])
IDs = paste(resultZ$Chromosome,resultZ$Begin,resultZ$End,sep="_")
dups = which(duplicated(IDs))
if(length(dups)>0) {resultZ = resultZ[-dups,]}
rownames(resultZ) <- paste(resultZ$Chromosome,resultZ$Begin,resultZ$End,sep="_")

annotZerone <- read.table(file.path(annotDir,"gencode.v34.annotation_V2.bed"),sep="\t")
colnames(annotZerone) <- c("chr","start","end","transcripts","Gene","strand")
annotZerone$ID = paste0(annotZerone$chr,"_",annotZerone$start,"_",annotZerone$end)
if(length(dups)>0){annotZerone = annotZerone[-dups,]}
rownames(annotZerone) <- annotZerone$ID
annotZerone$lengthPeak <- annotZerone$end - annotZerone$start

#Variables
annotCol <- c("Treatment","Exp")
pcaText <- TRUE
annotText <- "Sample"
hcText <- "Name"  ## column used in hierarchical clustering # change names from Sample_x to actual sample name
centering <- c("none","mean","median")[1]
##Hierarchical clustering
distHC <- c("distPearson","distCosine","euclidean","maximum","manhattan","canberra","binary","minkowski")[1]
methHC <- c("ward","ward.D","ward.D2","single","complete","average")[2]
##ConsClust
repsCC <- 1000
pItemCC <- 0.8
pFeatureCC <- 1
clusterAlgCC <- c("hc","pam","km","kmdist")[1]
distCC <- c("pearson","distCosine","euclidean","manhattan")[1]
innerLinkageCC <- c("ward","ward.D","ward.D2","single","complete","average")[2]
finalLinkageCC <- c("ward","ward.D","ward.D2","single","complete","average")[2]
## Heatmap
chRangeHM <-FALSE # Should be set to TRUE for expression data, FALSE for methylation data
hmColors <- colorRampPalette(c("royalblue","white","indianred1"))(256)
corColors <- colorRampPalette(c("indianred","white","forestgreen"))(256)
maxKHC <- 10
maxKCC <- 10

```


```{r, echo=FALSE}
annotZerone <- annotZerone[rownames(resultZ),]
```

```{r, echo=FALSE}
#Preparing raw count and normalized count matrices
rownames(resultZ) <- paste(resultZ$Chromosome,resultZ$Begin,resultZ$End,sep="_")
RZ <- (resultZ[,-c(1:3,grep(pattern = "input",colnames(resultZ)))]) #RZ matrix is the raw count matrix

#create RPKM matrix by normalizing per library size and peak length
RZ_RPKM <- (10^9*t(t((RZ+1)/annotZerone$lengthPeak)/colSums(RZ)))
```


```{r, echo=FALSE}
#Selecting samples for unsupervised analysis

RZ_sel <- RZ
RZ_sel <- (RZ_sel[,-grep(pattern = "GSK",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
RZ_sel <- (RZ_sel[,-grep(pattern = "MM468_1",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
RZ_sel <- (RZ_sel[,-grep(pattern = "spike",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
# RZ_sel <- (RZ_sel[,-grep(pattern = "_b_",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
dim(RZ_sel)
colnames(RZ_sel)
ExpName <- "MM468_K27_transcripts_10k_bulk"
dim(RZ_sel)
annotZerone_sel <- annotZerone; dim(annotZerone_sel)

exp.sd <- apply(RZ_sel, 1, sd)
sel2 <- order(exp.sd, decreasing = TRUE)[1:1000]
RZ_sel2 <- RZ_sel[sel2,]; dim(RZ_sel2)
annotZerone_sel2 <- annotZerone_sel[sel2,]

mat <- log2(RZ_RPKM[rownames(RZ_sel2),colnames(RZ_sel2)])
dim(mat)
hist(mat,breaks=100,main="Distribution of log2 pseudo IP")

```

```{r, echo=FALSE}
library(scTools)
  resSUBdir <- file.path(resdir1, "Correlation") ; if(!file.exists(resSUBdir)){dir.create(resSUBdir)}
  RDatadir <- file.path(resSUBdir,"RData") ; if(!file.exists(RDatadir)){dir.create(RDatadir)}
  annotSample = annot_sel <- data.frame(
    "Name" = colnames(mat),
    "Treatment"= gsub(".*_","",
                              gsub("J...|J..","",
                                   gsub("_K27","",colnames(mat)))),
    "Exp" =paste0("Exp_",gsub("_.*","",
                              gsub("MM468_","",
                                   gsub("_K27","",colnames(mat))))
    ),
    "day" = gsub("_.*","",
                 gsub(".*_J","",colnames(mat))) 
  )
  annotSample$Treatment[9:13] = c("5FUR","5FUR","5FUR","DMSO","DMSO")
  annotSample$Exp[9:13] = c("Exp_6","Exp_6","Exp_6","Exp_6","Exp_6")
  annotSample$day[9:13] = c(33,33,33,33,33)
  annot_sel =annotSample 
  anocol <- geco.annotToCol2(annotS=annot_sel[,annotCol],annotT=annot_sel,plotLegend=T,plotLegendFile=file.path(resSUBdir,"Annotation_legends.pdf"), categCol=NULL)
  
  control <- c("#E0E0E0","#BDBDBD","#757575","#474747")
  persister_color <- c("#DCEDC8","#9CCC65","#4CAF50","#009688")
  treated_UNC_color <- c("#2196F3", "#4DD0E1" )
  treated_GSKJ4_color <- c("#C2185B", "#EC407A" )
  res_color <- c("#FFEB3B","#FFC107","#FF9800","#FF5722")

  anocol[anocol[,1]=="mediumaquamarine",1] <- "#9CCC65"
  anocol[anocol[,1]=="plum1",1] <- "#BDBDBD"
  anocol[anocol[,2]=="#7FC97F",2] <- "#E7F0C3"
    anocol[anocol[,2]=="#BEAED4",2] <- "#F2A6A6"
    anocol[anocol[,2]=="#FDC086",2] <- "#FFEADB"
    
  save(annotSample,file=file.path(RDatadir,"annot.RData"))
  save(anocol,file=file.path(RDatadir,"anocol.RData"))
  save(mat,file=file.path(RDatadir,"mat.RData"))
```


##1.PCA
```{r, echo=FALSE}  
  ################################################################################
  ## PCA
  ################################################################################
  pca <- prcomp(t(mat),center=T,scale=F)

  save(pca,file=file.path(RDatadir,"pca.RData"))

  
  ## Barplot of standard deviations of each PC
  #pdf(file.path(resSUBdir,"PCA_stdev_barplot.pdf"), height=6, width=6)
  variances <- summary(pca)$importance["Proportion of Variance",]
  variances <- variances[1:(length(variances)-1)] # remove last PC as not informative
  barplot(variances, names.arg = names(variances), cex.names = 1, ylab = "Proportion of Variance", xlab="Principal components", las=2,col="royalblue",border="darkblue")
  #dev.off()
  
  ## Automatic 2D PCA plots all versus all (up to the PC number specified)
  # significant.PCs <- 4 # Can be determined from standard deviations plot
  significant.PCs <- ncol(pca$x)-2 ; if (significant.PCs > 6) {significant.PCs <- 6}
  combinations <- as.data.frame(combn(significant.PCs,2,simplify=TRUE))
  extendXlimLeft <- 5 # Extend left Xlim by x percent of min
  extendXlimRight <- 15 # Extend right Xlim by x percent of max
  extendYlimBottom <- 5 # Extend bottom Ylim by x percent of min
  extendYlimTop <- 5 # Extend top Ylim by x percent of max
  TextSize <- 0.4
 
  i=1
 png(file.path(resSUBdir,paste0(ExpName,sprintf("PCA%sVs%s_plot_2D_woText.png", combinations[1,i], combinations[2,i] ))), height=900,width=800,res=300)
    j=1
    xlimits <- c(-15,15)
      ylimits <- c(-10,10)
       plot(pca$x[,combinations[1,i]],pca$x[,combinations[2,i]],col=anocol[,j],xlab=paste(colnames(pca$x)[combinations[1,i]],100*variances[combinations[1,i]]),ylab=paste(colnames(pca$x)[combinations[2,i]],100*variances[combinations[2,i]]),cex=0.8,lwd=1.5, main=colnames(anocol)[j], pch=19, 
           xlim=xlimits,
           ylim=ylimits
      ) 
  dev.off() 
    
```

##2. Correlation plots  
```{r, echo=FALSE}
  ################################################################################
  ## Cor Clustering on all data
  ################################################################################

  hc_cor <- hclust(as.dist(1 - cor(mat)), method = "ward.D")
  mat.so <- mat[,hc_cor$order]

  
  if(chRangeHM){
    for(i in 1:nrow(mat.so)){
      mat.so[i,] <- geco.changeRange(mat.so[i,],newmin=0,newmax=1)
    }
  }
  
  png(file.path(resSUBdir,paste0(ExpName,"_Clustering_correlation_matrix.png")), height=1000,width=1000,res=150)
  geco.hclustAnnotHeatmapPlot(x=cor(mat.so),
                              hc=hc_cor,
                              hmColors=hmColors,
                              anocol=anocol[hc_cor$order,],#[,ncol(cc.col):1]
                              xpos=c(0.15,0.9,0.164,0.885),
                              ypos=c(0.1,0.5,0.5,0.6,0.62,0.95),
                              dendro.cex=0.5,
                              xlab.cex=0.5,
                              hmRowNames=TRUE,
                              hmRowNames.cex=0.01
  )
 dev.off()
 
  png(file.path(resSUBdir,paste0(ExpName,"_legend_clustering.png")), height=700,width=1000,res=150)
  par(mar = c(11, 8, 8, 8))
  lims <- seq(-1, 1, length.out = 200)
  image(matrix(lims, nc = 1), col = hmColors, axes = F, 
  xlab = paste0("min ",round(min(cor(mat.so)),2)," max ",max(cor(mat.so))))
  dev.off()
   
  geco.hclustAnnotHeatmapPlot(x=cor(mat.so),
                              hc=hc_cor,
                              hmColors=hmColors,
                              anocol=anocol[hc_cor$order,],#[,ncol(cc.col):1]
                              xpos=c(0.15,0.9,0.164,0.885),
                              ypos=c(0.1,0.5,0.5,0.6,0.62,0.95),
                              dendro.cex=0.5,
                              xlab.cex=0.5,
                              hmRowNames=TRUE,
                              hmRowNames.cex=0.01 
 )
  
 ################################################################################
  ## Corrplot package
  ################################################################################
  
  corrplot(cor(mat),order="hclust",tl.col = "black",tl.cex=0.2,cl.cex=0.3,col=hmColors)
```
  
##4. Diff analysis on IP signal only
```{r, echo=FALSE}
library(geco.supervised)
library(ggplot2)
library(RColorBrewer)
library(xtable)
library(geco.ChIPseq)
library(WriteXLS)
library(gtools)
library(limma)
library(edgeR)
options(width=180)


qval.th <- 0.1
logFC.th =2

resdir <- file.path(resdir1, "Supervised/");if(!file.exists(resdir)){dir.create(resdir)}
tabdir <- file.path(resdir,"Tables");if(!file.exists(tabdir)){dir.create(tabdir)}
plotdir <- file.path(resdir,"Plots");if(!file.exists(plotdir)){dir.create(plotdir)}
EnrichDir <- file.path(tabdir,"Gene_set_analysis");if(!file.exists(EnrichDir)){dir.create(EnrichDir)}
RDataSupdir <-  file.path(resdir,"RData");if(!file.exists(RDataSupdir)){dir.create(RDataSupdir)}

library(ChromSCape)
data("hg38.MSIG.gs")
data("hg38.MSIG.ls")

myrefs <- list(
DMSO6= annot_sel[which(annot_sel$Treatment=="DMSO" & annot_sel$Exp=="Exp_6"),"Name"]
)
mygps <- list(
X5FU6 = annot_sel[which(annot_sel$Treatment=="5FUR" & annot_sel$Exp %in% c("Exp_6")),"Name"]
)

refs <- names(myrefs)
groups <- names(mygps)

rownames(annot_sel) <- annot_sel$Name

#selection of peaks with at least a log2 RPKM of 1 in one sample
sel <- which(apply(RZ_RPKM[,colnames(RZ_sel2)],1,max)>1)

annotZerone. = annotZerone[,c("chr","start","end","ID")]
res <- geco.ChIPseqCompareLimma(mat=RZ,
                                metadata = annot_sel,
                                ref=myrefs,
                                groups=mygps,
                                featureTab=annotZerone.[rownames(RZ),]
)
res = cbind(res,annotZerone[,c("lengthPeak","Gene","transcripts")])

save(res, file=file.path(RDataSupdir,paste("Supervised_res_Limma_object.Rdata",sep="")))
  under_res = res %>% dplyr::filter(log2FC.X5FU6 < -logFC.th & qval.X5FU6 < qval.th) %>% 
    dplyr::arrange(qval.X5FU6) %>% select(Gene,log2FC.X5FU6,qval.X5FU6,ID,lengthPeak,transcripts)
over_res = res %>% dplyr::filter(log2FC.X5FU6 > logFC.th & qval.X5FU6 < qval.th) %>% 
    dplyr::arrange(qval.X5FU6) %>% select(Gene,log2FC.X5FU6,qval.X5FU6,ID,lengthPeak,transcripts)

WriteXLS(c("over_res","under_res","res"), ExcelFileName =   file.path(tabdir,"Differential_analysis_Limma.xlsx"), SheetNames = c(paste0("Over_",logFC.th,"_",qval.th,"_n",nrow(over_res)),paste0("Under_-",logFC.th,"_",qval.th,"_n",nrow(under_res)),"All"), perl = "perl", verbose = FALSE, row.names = FALSE, col.names = TRUE, AdjWidth = T, AutoFilter = TRUE, BoldHeaderRow = TRUE, na = "", FreezeRow = 1, FreezeCol = 1)

summaryTab <- geco.summaryChIPseqCompareLimma(restab=res,
										ref=myrefs,
							 		 groups=mygps,
  							 		 sign.th=0.1,
							 		 plotdir=plotdir,
								     fc.th=2,
										volcano_dotcex = 1,
										volcano_dot_alpha = 0.2,
										volcano_add_xlim = 2,
										display_top_names=0
										)
				 
summaryTab <- data.frame("."=row.names(summaryTab), summaryTab, check.names=FALSE)		
write.table(summaryTab, file.path(tabdir,"Number_differentially_bound_peaks_per_group.csv"),row.names=F,quote=F,sep=";"  )

summaryTab <- geco.summaryChIPseqCompareLimma(restab=res,
										ref=myrefs,
							 		 groups=mygps,
  							 		 sign.th=0.1,
							 		 plotdir=plotdir,
								     fc.th=2,
										volcano_dotcex = 1,
										volcano_dot_alpha = 0.2,
										volcano_add_xlim = 2,
										display_top_names=10
										)
 ```
```{r,include=FALSE}
par(mar=c(5,5,0,0))
``` 

##5. Enrichment analysis

```{r, echo=FALSE}
###Enrichment analysiss

data("hg38.GeneTSS")
GencodeGene = hg38.GeneTSS 
possibleGenes <- c(as.character(unique(GencodeGene$gene)),"ELFN2")

resTSS <- res #[res$peak_affectation=="tss_pc",]
reflist <- possibleGenes
annotbase<- "MSigDB" 
MSIG.gs = hg38.MSIG.gs
enrichON <- MSIG.gs$Gene.Set[MSIG.gs$Class %in% c("c2_curated","c6_oncogenic","hallmark")] #remove unwanted gene lists to gain statistical power
MSIG.ls = hg38.MSIG.ls
database <- MSIG.ls[enrichON]
resTSS$Gene = as.character(resTSS$Gene)
resTSS = resTSS[which(resTSS$Gene %in% hg38.GeneTSS$gene),]

for(i in 1:length(groups)) 	{
    if (length(refs)>1) {ref <- refs[i]} else {ref <- refs[1]}
    print(paste0("Processing ",groups[i], " vs ", ref, " _ ",annotbase ))

      	signific <- which(resTSS[,paste("qval",groups[i],sep=".")] <= qval.th)
	      over     <- which(resTSS[,paste("qval",groups[i],sep=".")] <= qval.th & resTSS[,paste("log2FC",groups[i],sep=".")] > logFC.th)
	      under    <- which(resTSS[,paste("qval",groups[i],sep=".")] <= qval.th & resTSS[,paste("log2FC",groups[i],sep=".")] < (-logFC.th))
   
      if(length(over)){
        enrich.test <- geco.enrichmentTest(gene.sets=database,
                                           mylist=unique(unlist(strsplit(resTSS$Gene[over],split=","))),
                                           possibleIds=reflist)
        enrich.test <- data.frame(Gene_set_name=rownames(enrich.test), enrich.test, check.names=FALSE)
        enrich.test <- merge( subset(MSIG.gs, select=-Genes), enrich.test, by.x="Gene.Set", by.y="Gene_set_name", all.y=TRUE, sort=FALSE ) ## Get class of gene set
        enrich.test <- enrich.test[order(enrich.test$`p-value`),]
        ind <- which(enrich.test$`q-value`<= qval.th);if(!length(ind)){ind <- 1:20}
        Overexpressed  <- enrich.test[ind,]
        png(file.path(EnrichDir,paste0(groups[i],"_vs_",ref,"_qval",qval.th,"_logFC",logFC.th,"_enriched_peaks_enrichment.png")), height=1000,width=1000,res=150)
        par(mar = c(13, 4, 2, 2) + 1)
        barplot(-log10(enrich.test$`q-value`[1:10]),las=2,names.arg = enrich.test$Gene.Set[1:10],cex.names = 0.5,ylab="-log10 q-value",main=paste0("Top 10 gene list of ",length(ind)," lists"))
        dev.off()
        }
    
      if(length(under)){
        enrich.test <- geco.enrichmentTest(gene.sets=database,mylist=unique(unlist(strsplit(resTSS$Gene[under],split=","))),possibleIds=reflist)
        enrich.test <- data.frame(Gene_set_name=rownames(enrich.test), enrich.test, check.names=FALSE)
        enrich.test <- merge( subset(MSIG.gs, select=-Genes), enrich.test, by.x="Gene.Set", by.y="Gene_set_name", all.y=TRUE, sort=FALSE ) ## Get class of gene set
        enrich.test <- enrich.test[order(enrich.test$`p-value`),]
        ind <- which(enrich.test$`q-value`<= qval.th);if(!length(ind)){ind <- 1:20}
        Underexpressed <- enrich.test[ind,]	
        
        png(file.path(EnrichDir,paste0(groups[i],"_vs_",ref,"_qval",qval.th,"_logFC",logFC.th,"_depleted_peaks_enrichment.png")), height=1000,width=1000,res=150)
        par(mar = c(13, 4, 2, 2) + 1)
        barplot(-log10(enrich.test$`q-value`[1:10]),las=2,names.arg = enrich.test$Gene.Set[1:10],cex.names = 0.5,ylab="-log10 q-value",main=paste0("Top 10 gene list of ",length(ind)," lists") )
        dev.off()	}
    
    WriteXLS(c("Overexpressed", "Underexpressed"), ExcelFileName =   file.path(EnrichDir,paste0("Enrichment_test_",groups[i],"_vs_",ref,"_",annotbase,"_",qval.th,"_",logFC.th,".xlsx")), SheetNames = c( paste0("Enriched_in_", groups[i]), paste0("Depleted_in_", groups[i])), perl = "perl", verbose = FALSE, row.names = FALSE, col.names = TRUE, AdjWidth = TRUE, AutoFilter = TRUE, BoldHeaderRow = TRUE, na = "", FreezeRow = 1, FreezeCol = 1)

    }

```


# 5. Comparison differential regions bulk with regions H3K4me3
```{r, echo=FALSE}
diff_regions = res$ID[which(abs(res$log2FC.X5FU6)>logFC.th & res$qval.X5FU6 < qval.th)]

q = quantile(RZ_RPKM[,"MM468BC_DMSO_D33_C02_K4"], 0.5)

a = length(which(RZ_RPKM[diff_regions,"MM468BC_DMSO_D33_C02_K4"] < q))
c = length(which(RZ_RPKM[,"MM468BC_DMSO_D33_C02_K4"] < q))
b = length(diff_regions)
d = nrow(RZ_RPKM)
x = matrix(c(a,b-a,c,d-c), ncol=2)
x
fisher.test(x,alternative = "greater")

mat_5FU6_sc = read.table("../../../../../Documents/Data/results/Matrices_hg38/MM468_K4_transcripts_10k_1000/MM468_5FU6_D60_K4.tsv", header = T,sep = "\t",row.names = 1)
mat_DMSO6_sc = read.table("../../../../../Documents/Data/results/Matrices_hg38/MM468_K4_transcripts_10k_1000/MM468_Ipop_DMSO6_K4me3.tsv", header = T,sep = "\t",row.names = 1)

mat <- matrix(c(RZ_RPKM[-which(rownames(RZ_RPKM) %in% diff_regions),"MM468BC_DMSO_D33_C02_K4"],
                RZ_RPKM[diff_regions,"MM468BC_DMSO_D33_C02_K4"],
                rep(0,nrow(RZ_RPKM) - 2*length(diff_regions))), ncol = 2)
df <- data.frame(values = mat[1:nrow(RZ_RPKM)],
                 vars = rep(c("Not differential","Differential"), times = c(nrow(RZ_RPKM)-length(diff_regions),
                                                        length(diff_regions))))

## draw the boxplot
boxplot(log2(values+1) ~ vars, data = df, xlab= "", ylab="log2(Count)", main = "DMSO")


mat <- matrix(c(RZ_RPKM[-which(rownames(RZ_RPKM) %in% diff_regions),"MM468BC_5FU_D33_E02_K4"],
                RZ_RPKM[diff_regions,"MM468BC_5FU_D33_E02_K4"],
                rep(0,nrow(RZ_RPKM) - 2*length(diff_regions))), ncol = 2)
df <- data.frame(values = mat[1:nrow(RZ_RPKM)],
                 vars = rep(c("Not differential","Differential"), times = c(nrow(RZ_RPKM)-length(diff_regions),
                                                        length(diff_regions))))

## draw the boxplot
boxplot(log2(values+1) ~ vars, data = df, xlab= "", ylab="log2(Count)", main = "5FU")

# Difference
mat <- matrix(c(log2(RZ_RPKM[-which(rownames(RZ_RPKM) %in% diff_regions),"MM468BC_5FU_D33_E02_K4"] /
                  RZ_RPKM[-which(rownames(RZ_RPKM) %in% diff_regions),"MM468BC_DMSO_D33_C02_K4"]),
                log2(RZ_RPKM[diff_regions,"MM468BC_5FU_D33_E02_K4"] / 
                  RZ_RPKM[diff_regions,"MM468BC_DMSO_D33_C02_K4"]),
                rep(0,nrow(RZ_RPKM) - 2*length(diff_regions))), ncol = 2)
df <- data.frame(values = mat[1:nrow(RZ_RPKM)],
                 vars = rep(c("Not differential","Differential"), times = c(nrow(RZ_RPKM)-length(diff_regions),
                                                        length(diff_regions))))

## draw the boxplot
boxplot(values ~ vars, data = df, xlab= "", ylab="log2(5FU/DMSO)", main = "5FU / DMSO")

```


