---
title: "MM468 bulk peaks K27"
author: "CÃ©line Vallot"
date: "05/06/2019"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
library(scran)
library(corrplot)
library(devtools)
library(pheatmap)
library(limma)
library(scater)
library(statmod)
library(Matrix)
library(dplyr)
library(geco.utils)
library(geco.visu)
library(geco.unsupervised)
library(geco.ChIPseq)
library(Rtsne)
library(RColorBrewer)
library(colorRamps)
library(dplyr)
library(tidyr)
library(viridis)
library(wesanderson)
library(knitr)
library(ChromSCape)
options(stringsAsFactors = F)
```

```{r, echo=FALSE}
#Directories
dataDir <- "K27_peaks_K27/07csv/"
annotDir <- "peaks/"
resdir1 <- "K27_peaks_K27/results/"

#Count matrices
resultZ <- read.csv(file.path(dataDir,"results_Zerone_merge_hg38_K27_annot.csv"))
rownames(resultZ) <- paste(resultZ$Chromosome,resultZ$Begin,resultZ$End,sep="_")

#Sample annotation
annotSample <- read.csv(file.path(dataDir,"BowtieStats_annotated.csv"),sep=";")
annotSample = annotSample[,1:2]
colnames(annotSample) = c("Treatment","Exp")
rownames(annotSample) = gsub("_K27","",annotSample$Treatment)

#Variables
annotCol <- c("Treatment","Exp")
pcaText <- TRUE
annotText <- "Sample"
hcText <- "Name"  ## column used in hierarchical clustering # change names from Sample_x to actual sample name
centering <- c("none","mean","median")[1]
##Hierarchical clustering
distHC <- c("distPearson","distCosine","euclidean","maximum","manhattan","canberra","binary","minkowski")[1]
methHC <- c("ward","ward.D","ward.D2","single","complete","average")[2]
##ConsClust
repsCC <- 1000
pItemCC <- 0.8
pFeatureCC <- 1
clusterAlgCC <- c("hc","pam","km","kmdist")[1]
distCC <- c("pearson","distCosine","euclidean","manhattan")[1]
innerLinkageCC <- c("ward","ward.D","ward.D2","single","complete","average")[2]
finalLinkageCC <- c("ward","ward.D","ward.D2","single","complete","average")[2]
## Heatmap
chRangeHM <-FALSE # Should be set to TRUE for expression data, FALSE for methylation data
hmColors <- colorRampPalette(c("royalblue","white","indianred1"))(256)
corColors <- colorRampPalette(c("indianred","white","forestgreen"))(256)

maxKHC <- 10
maxKCC <- 10

```


```{r, echo=FALSE}

#Annotation of peaks with Gencode info, K4me3 peaks and enhancers of breast cancer
annotZerone <- read.table(file.path(annotDir,"MM468_peaks_K27.bed"),sep="\t")
colnames(annotZerone) <- c("chr","start","end","ID")
annotZerone$ID = paste0(annotZerone$chr,"_",annotZerone$start,"_",annotZerone$end)
rownames(annotZerone) <- annotZerone$ID
annotZerone$lengthPeak <- annotZerone$end - annotZerone$start

annotZerone_genebody <- read.table(file.path(annotDir,"MM468_peaks_K27.pc.genebody.bed"),sep="\t")
annotZerone_genebody <- annotZerone_genebody[,-c(8:10)]
colnames(annotZerone_genebody) <- c("chr","start","end","ID","chrG","startG","endG","Gene")
annotZerone_genebody <- aggregate(Gene ~ ID, data = annotZerone_genebody, paste, collapse = ",")
annotZerone$genebody <- ""
annotZerone$genebody <- annotZerone_genebody$Gene[match(annotZerone$ID,annotZerone_genebody$ID)]

annotZerone_genebody_all <- read.table(file.path(annotDir,"MM468_peaks_K27.all.genebody.bed"),sep="\t")
annotZerone_genebody_all <- annotZerone_genebody_all[,-c(8:10)]
colnames(annotZerone_genebody_all) <- c("chr","start","end","ID","chrG","startG","endG","Gene")
annotZerone_genebody_all <- aggregate(Gene ~ ID, data = annotZerone_genebody_all, paste, collapse = ",")
annotZerone$genebody_all <- ""
annotZerone$genebody_all <- annotZerone_genebody_all$Gene[match(annotZerone$ID,annotZerone_genebody_all$ID)]

annotZerone_tss_i <- read.table(file.path(annotDir,"MM468_peaks_K27.distance.pc.transcript.tss.bed"),sep="\t")[,-c(5:7)]
colnames(annotZerone_tss_i) <- c("chr","start","end","ID","Gene","distance")
annotZerone_tss_i <- annotZerone_tss_i[!duplicated(annotZerone_tss_i),]
annotZerone_tss <- aggregate(Gene ~ ID, data = annotZerone_tss_i, paste, collapse = ",")
annotZerone_tss$distance <- aggregate(distance ~ ID, data = annotZerone_tss_i,min )[,2]
annotZerone$transcript <- ""
annotZerone$transcript <- annotZerone_tss$Gene[match(annotZerone$ID,annotZerone_tss$ID)]
annotZerone$distance <- ""
annotZerone$distance <- annotZerone_tss$distance[match(annotZerone$ID,annotZerone_tss$ID)]

annotZerone_tss_all_i <- read.table(file.path(annotDir,"MM468_peaks_K27.distance.all.transcript.tss.bed"),sep="\t")[,-c(5:7)]
colnames(annotZerone_tss_all_i) <- c("chr","start","end","ID","Gene","distance")
annotZerone_tss_all_i <- annotZerone_tss_all_i[!duplicated(annotZerone_tss_all_i),]
annotZerone_tss_all <- aggregate(Gene ~ ID, data = annotZerone_tss_all_i, paste, collapse = ",")
annotZerone_tss_all$distance <- aggregate(distance ~ ID, data = annotZerone_tss_all_i,min )[,2]
annotZerone$transcript_all <- ""
annotZerone$transcript_all <- annotZerone_tss_all$Gene[match(annotZerone$ID,annotZerone_tss_all$ID)]
annotZerone$distance_all <- ""
annotZerone$distance_all <- annotZerone_tss_all$distance[match(annotZerone$ID,annotZerone_tss_all$ID)]

annotZerone$peak_affectation <- "intergenic"
annotZerone$peak_affectation[annotZerone$distance<1000] <- "tss_pc"
annotZerone$peak_affectation[annotZerone$distance>1000 & !is.na(annotZerone$genebody)] <- "genebody_pc"
annotZerone$peak_affectation[annotZerone$distance>1000 & is.na(annotZerone$genebody) & annotZerone$distance_all<1000] <- "tss_other"
annotZerone$peak_affectation[annotZerone$distance>1000 & is.na(annotZerone$genebody) & annotZerone$distance_all>1000 & !is.na(annotZerone$genebody_all)] <- "genebody_other"

annotZerone$gene_affectation <- "intergenic"
annotZerone$gene_affectation[annotZerone$distance<1000] <- annotZerone$transcript[annotZerone$distance<1000]
annotZerone$gene_affectation[annotZerone$distance>1000 & !is.na(annotZerone$genebody)] <- annotZerone$genebody[annotZerone$distance>1000 & !is.na(annotZerone$genebody)]
annotZerone$gene_affectation[annotZerone$distance>1000 & is.na(annotZerone$genebody) & annotZerone$distance_all<1000] <- annotZerone$transcript_all[annotZerone$distance>1000 & is.na(annotZerone$genebody) & annotZerone$distance_all<1000]
annotZerone$gene_affectation[annotZerone$distance>1000 & is.na(annotZerone$genebody) & annotZerone$distance_all>1000 & !is.na(annotZerone$genebody_all)] <- annotZerone$genebody_all[annotZerone$distance>1000 & is.na(annotZerone$genebody) & annotZerone$distance_all>1000 & !is.na(annotZerone$genebody_all)]

table(annotZerone$peak_affectation)

annotZerone_K4 <- read.table(file.path(annotDir,"MM468_peaks_K27.K4.bed"),sep="\t")
colnames(annotZerone_K4) <- c("chr","start","end","ID","chrG","startG","endG")
annotZerone_K4$ID_K4 <- paste(annotZerone_K4$chr,annotZerone_K4$start,annotZerone_K4$end,sep="_")
annotZerone$K4_ID <- annotZerone_K4$ID_K4[match(annotZerone$ID,annotZerone_K4$ID)]
annotZerone$K4 <- !is.na(annotZerone$K4_ID)

table(annotZerone$peak_affectation, annotZerone$K4)

annotZerone$bivalent <- (annotZerone$peak_affectation %in% c("tss_pc") & annotZerone$K4 )

annotZerone_enhancer <- read.table(file.path(annotDir,"MM468_peaks_K27_enhancers.bed"),sep="\t",row.names = 1)
annotZerone_enhancer <- annotZerone_enhancer[rownames(annotZerone),]
annotZerone$enhancer <- annotZerone_enhancer$V8==0
sel <- (annotZerone$peak_affectation=="intergenic" & annotZerone$enhancer)
annotZerone$peak_affectation[sel] <- "intergenic_enhancer"

annotZerone$peak_affectation <- factor(annotZerone$peak_affectation,levels=c("intergenic","intergenic_enhancer","tss_pc","genebody_pc","tss_other","genebody_other"))
# #missing coordinates of enhancer to add

annotZerone <- annotZerone[rownames(resultZ),]


```

```{r, echo=FALSE}
#Preparing raw count and normalized count matrices
rownames(resultZ) <- paste(resultZ$Chromosome,resultZ$Begin,resultZ$End,sep="_")
RZ <- (resultZ[,-c(1:3,grep(pattern = "input",colnames(resultZ)))]) #RZ matrix is the raw count matrix

#create RPKM matrix by normalizing per library size and peak length
RZ_RPKM <- (10^9*t(t((RZ+1)/annotZerone$lengthPeak)/colSums(RZ)))
```


```{r, echo=FALSE}
#Selecting samples for unsupervised analysis

RZ_sel <- RZ
RZ_sel <- (RZ_sel[,-grep(pattern = "_C1_",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
RZ_sel <- (RZ_sel[,-grep(pattern = "_a_",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
RZ_sel <- (RZ_sel[,-grep(pattern = "K4",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
RZ_sel <- (RZ_sel[,-grep(pattern = "231",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
RZ_sel <- (RZ_sel[,-grep(pattern = "436",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
RZ_sel <- (RZ_sel[,-grep(pattern = "GSK",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
RZ_sel <- (RZ_sel[,-grep(pattern = "MM468_1",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
RZ_sel <- (RZ_sel[,-grep(pattern = "spike",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
RZ_sel <- (RZ_sel[,-grep(pattern = "_b_",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
# RZ_sel <- (RZ_sel[,-grep(pattern = "dosetest",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
dim(RZ_sel)
colnames(RZ_sel)
ExpName <- "MM468_K27_exp_2_3_5_bulk"

dim(RZ_sel)
annotZerone_sel <- annotZerone; dim(annotZerone_sel)

exp.sd <- apply(RZ_sel, 1, sd)
sel2 <- order(exp.sd, decreasing = TRUE)[1:1000]
RZ_sel2 <- RZ_sel[sel2,]; dim(RZ_sel2)
annotZerone_sel2 <- annotZerone_sel[sel2,]

mat <- log2(RZ_RPKM[rownames(RZ_sel2),colnames(RZ_sel2)])
dim(mat)
hist(mat,breaks=100,main="Distribution of log2 pseudo IP")

```

```{r, echo=FALSE}
library(scTools)
  resSUBdir <- file.path(resdir1, "Correlation") ; if(!file.exists(resSUBdir)){dir.create(resSUBdir)}
  RDatadir <- file.path(resSUBdir,"RData") ; if(!file.exists(RDatadir)){dir.create(RDatadir)}
  annotSample <- data.frame(
    "Name" = colnames(mat),
    "Treatment"= gsub(".*_","",
                              gsub("J...|J..","",
                                   gsub("_K27","",colnames(mat)))),
    "Exp" =paste0("Exp_",gsub("_.*","",
                              gsub("MM468_","",
                                   gsub("_K27","",colnames(mat))))
    ),
    "day" = gsub("_.*","",
                 gsub(".*_J","",colnames(mat))) 
  )
  
  annotSample[9,"Treatment"] = "DMSO" #specific
  annotSample[9,"Exp"] = "Exp_3" #specific
  annotSample[9,"day"] = 3 #specific
  annotSample = annotSample[order(annotSample$Treatment),]
  annot_sel = annotSample
  
  
  anocol <- geco.annotToCol2(annotS=annot_sel[,annotCol],
                             annotT=annot_sel,plotLegend=T,
                             plotLegendFile=NULL, categCol=NULL)
  
  control <- c("#E0E0E0","#BDBDBD","#757575","#474747")
  persister_color <- c("#DCEDC8","#9CCC65","#4CAF50","#009688")
  treated_UNC_color <- c("#2196F3", "#4DD0E1" )
  treated_GSKJ4_color <- c("#C2185B", "#EC407A" )
  res_color <- c("#FFEB3B","#FFC107","#FF9800","#FF5722")

  anocol[anocol[,1]=="mediumaquamarine",1] <- "#9CCC65"
  anocol[anocol[,1]=="plum1",1] <- "#BDBDBD"
  anocol[anocol[,2]=="#7FC97F",2] <- "#E7F0C3"
    anocol[anocol[,2]=="#BEAED4",2] <- "#F2A6A6"
    anocol[anocol[,2]=="#FDC086",2] <- "#FFEADB"


  # if(centering=="mean")	mat <- mat-apply(mat,1,mean)
  # if(centering=="median")	mat <- mat-apply(mat,1,median)
  
  save(annotSample,file=file.path(RDatadir,"annot.RData"))
  save(anocol,file=file.path(RDatadir,"anocol.RData"))
  save(mat,file=file.path(RDatadir,"mat.RData"))
```


##1.PCA
```{r, echo=FALSE}  
  ################################################################################
  ## PCA
  ################################################################################
  pca <- prcomp(t(mat),center=T,scale=F)

  save(pca,file=file.path(RDatadir,"pca.RData"))

  
  ## Barplot of standard deviations of each PC
  variances <- summary(pca)$importance["Proportion of Variance",]
  variances <- variances[1:(length(variances)-1)] # remove last PC as not informative
  barplot(variances, names.arg = names(variances), cex.names = 1,
          ylab = "Proportion of Variance", xlab="Principal components",
          las=2,col="royalblue",border="darkblue")
  
  ## Automatic 2D PCA plots all versus all (up to the PC number specified)
  significant.PCs <- ncol(pca$x)-2 ; if (significant.PCs > 6) {significant.PCs <- 6}
  combinations <- as.data.frame(combn(significant.PCs,2,simplify=TRUE))
  extendXlimLeft <- 5 # Extend left Xlim by x percent of min
  extendXlimRight <- 15 # Extend right Xlim by x percent of max
  extendYlimBottom <- 5 # Extend bottom Ylim by x percent of min
  extendYlimTop <- 5 # Extend top Ylim by x percent of max
  TextSize <- 0.4
 
  i=1
 png(file.path(resSUBdir,paste0(ExpName,sprintf("PCA%sVs%s_plot_2D_woText.png", combinations[1,i], combinations[2,i] ))), height=900,width=800,res=300)
    j=1
    xlimits <- c(-15,15)
      ylimits <- c(-10,10)
       plot(pca$x[,combinations[1,i]],pca$x[,combinations[2,i]],col=anocol[,j],xlab=paste(colnames(pca$x)[combinations[1,i]],100*variances[combinations[1,i]]),ylab=paste(colnames(pca$x)[combinations[2,i]],100*variances[combinations[2,i]]),cex=0.8,lwd=1.5, main=colnames(anocol)[j], pch=19, 
           xlim=xlimits,
           ylim=ylimits
      ) 
  dev.off() 
    
```

##2. Correlation plots  
```{r, echo=FALSE}
  ################################################################################
  ## Cor Clustering on all data
  ################################################################################
 
  colnames(mat) <- annot_sel$Name[match(colnames(mat),annot_sel$Name)]
  hc_cor <- hclust(as.dist(1 - cor(mat)), method = "ward.D")
  mat.so <- mat[,hc_cor$order]

  
  if(chRangeHM){
    for(i in 1:nrow(mat.so)){
      mat.so[i,] <- geco.changeRange(mat.so[i,],newmin=0,newmax=1)
    }
  }
  
  png(file.path(resSUBdir,paste0(ExpName,"_Clustering_correlation_matrix.png")), height=1000,width=1000,res=150)
  geco.hclustAnnotHeatmapPlot(x=cor(mat.so),
                              hc=hc_cor,
                              hmColors=hmColors,
                              anocol=anocol[hc_cor$order,],#[,ncol(cc.col):1]
                              xpos=c(0.15,0.9,0.164,0.885),
                              ypos=c(0.1,0.5,0.5,0.6,0.62,0.95),
                              dendro.cex=0.5,
                              xlab.cex=0.5,
                              hmRowNames=TRUE,
                              hmRowNames.cex=0.01
  )
 dev.off()
 
  png(file.path(resSUBdir,paste0(ExpName,"_legend_clustering.png")), height=700,width=1000,res=150)
  par(mar = c(11, 8, 8, 8))
  lims <- seq(-1, 1, length.out = 200)
  image(matrix(lims, nc = 1), col = hmColors, axes = F, 
  xlab = paste0("min ",round(min(cor(mat.so)),2)," max ",max(cor(mat.so))))
  dev.off()
   
  save(hc_cor,file=file.path(RDatadir,"hc_cor.RData"))

  
  geco.hclustAnnotHeatmapPlot(x=cor(mat.so),
                              hc=hc_cor,
                              hmColors=hmColors,
                              anocol=anocol[hc_cor$order,],#[,ncol(cc.col):1]
                              xpos=c(0.15,0.9,0.164,0.885),
                              ypos=c(0.1,0.5,0.5,0.6,0.62,0.95),
                              dendro.cex=0.5,
                              xlab.cex=0.5,
                              hmRowNames=TRUE,
                              hmRowNames.cex=0.01 
 )
  
 ################################################################################
  ## Corrplot package
  ################################################################################
  
  corrplot(cor(mat),order="hclust",tl.col = "black",tl.cex=0.2,cl.cex=0.3,col=hmColors)
```
  
##4. Diff analysis on IP signal only
```{r, echo=FALSE}
library(geco.supervised)
library(ggplot2)
library(RColorBrewer)
library(xtable)
library(geco.ChIPseq)
library(WriteXLS)
library(gtools)
library(limma)
library(edgeR)
options(width=180)

resdir <- file.path(resdir1, "Supervised/");if(!file.exists(resdir)){dir.create(resdir)}
tabdir <- file.path(resdir,"Tables");if(!file.exists(tabdir)){dir.create(tabdir)}
plotdir <- file.path(resdir,"Plots");if(!file.exists(plotdir)){dir.create(plotdir)}
EnrichDir <- file.path(tabdir,"Gene_set_analysis");if(!file.exists(EnrichDir)){dir.create(EnrichDir)}
RDataSupdir <-  file.path(resdir,"RData");if(!file.exists(RDataSupdir)){dir.create(RDataSupdir)}

# Change FC and Qval
qval.th <- 0.1
logFC.th=2 #FC of 3

library(ChromSCape)
data("hg38.MSIG.gs")
data("hg38.MSIG.ls")

myrefs <- list(
DMSO= annot_sel[which(annot_sel$Treatment=="DMSO"),"Name"]
)
mygps <- list(
X5FU2_3_5 = annot_sel[which(annot_sel$Treatment=="5FUR" & annot_sel$Exp %in% c("Exp_2","Exp_3","Exp_5")),"Name"]
)

refs <- names(myrefs)
groups <- names(mygps)

rownames(annot_sel) <- annot_sel$Name

#selection of peaks with at least a log2 RPKM of 1 in one sample
sel <- which(apply(RZ_RPKM[,colnames(RZ_sel2)],1,max)>1)
RZ_sel = RZ_sel[,annot_sel$Name]

res <- geco.ChIPseqCompareLimma(mat=RZ_sel,
                                metadata =annot_sel,
                                ref=myrefs,
                                groups=mygps,
                                featureTab=annotZerone[rownames(RZ_sel),]
)

under_res = res %>% dplyr::filter(log2FC.X5FU2_3_5 < -logFC.th & qval.X5FU2_3_5 < qval.th) %>% 
    dplyr::arrange(qval.X5FU2_3_5) %>% select(gene_affectation,log2FC.X5FU2_3_5,qval.X5FU2_3_5,ID,bivalent,lengthPeak)
over_res = res %>% dplyr::filter(log2FC.X5FU2_3_5 > logFC.th & qval.X5FU2_3_5 < qval.th) %>% 
    dplyr::arrange(qval.X5FU2_3_5) %>% select(gene_affectation,log2FC.X5FU2_3_5,qval.X5FU2_3_5,ID,bivalent,lengthPeak)

WriteXLS(c("over_res","under_res","res"), ExcelFileName =   file.path(tabdir,"Differential_analysis_Limma.xlsx"), SheetNames = c(paste0("Over_",logFC.th,"_",qval.th,"_n",nrow(over_res)),paste0("Under_-",logFC.th,"_",qval.th,"_n",nrow(under_res)),"All"), perl = "perl", verbose = FALSE, row.names = FALSE, col.names = TRUE, AdjWidth = T, AutoFilter = TRUE, BoldHeaderRow = TRUE, na = "", FreezeRow = 1, FreezeCol = 1)
 
save(res, file=file.path(RDataSupdir,paste("Supervised_res_Limma_object.Rdata",sep="")))
  

summaryTab <- geco.summaryChIPseqCompareLimma(
  restab=res,
  ref=myrefs,
  groups=mygps,
  sign.th=0.1,
  plotdir=plotdir,
  fc.th=2,
  volcano_dotcex = 1,
  volcano_dot_alpha = 0.2,
  volcano_add_xlim = 1
)
									 
summaryTab <- data.frame("."=row.names(summaryTab), summaryTab, check.names=FALSE)		
write.table(summaryTab, file.path(tabdir,"Number_differentially_bound_peaks_per_group.csv"),row.names=F,quote=F,sep=";"  )

library(kableExtra)

col = data.frame("col"=c("#999999ff","#dededeff","#3b9ab2ff","#78b7c5ff","#ebcc2aff","#e1af00ff"))
sig_diff <- (abs(res$log2FC.X5FU2_3_5)> logFC.th & res$qval.X5FU2_3_5 < qval.th)
sig_under <- (res$log2FC.X5FU2_3_5< -logFC.th & res$qval.X5FU2_3_5 < qval.th)
sig_over <- (res$log2FC.X5FU2_3_5> logFC.th & res$qval.X5FU2_3_5 < qval.th)
l = list("Diff" = sig_diff, "Under" = sig_under, "Over" = sig_over)
for(i in c("Diff","Under","Over")){
    sig = l[[i]]
    png(file.path(plotdir,paste0("enrichment_HQ_",i,".png")),width = 1500,height = 1500,res=300)
    sum_sig = sum(table(res$peak_affectation[sig]))
    sum_all = sum(table(res$peak_affectation))
    cbind(as.data.frame(log2((table(res$peak_affectation[sig])/sum_sig) / (table(res$peak_affectation)/sum_all))),col) %>% ggplot() + geom_bar(aes(x=Var1, y=Freq,fill= Var1) , width = 0.5,stat="identity") + ggtitle(paste0(i," enriched H3K27me3 peaks")) +
        theme_classic() + ggtitle(paste0(i," enriched H3K27me3 peaks")) + scale_fill_manual(values =col$col)  +
        theme(axis.text.x = element_text(angle = 90, vjust =1), legend.position = "None") + ylab("Log2 Enrichment")
    dev.off()
}

#Enrichment
fisher <- function(a,b,c,d){
  data <- matrix(c(a,b,c,d),ncol=2)
  c(p = fisher.test(data)$p.value)
}

sig = l[["Over"]]
sum_sig = sum(table(res$peak_affectation[sig]))
sum_all = sum(table(res$peak_affectation))

tab = res[sig,] %>% count(peak_affectation,.drop=F) %>% left_join(res %>% count(peak_affectation,.drop=F), by = "peak_affectation") %>% mutate(sum_sig =sum_sig, sum_all =sum_all, Freq = log2( (n.x/sum_sig) / (n.y/sum_all) ),
color = col$col) %>% rowwise() %>% mutate(p=fisher(n.x,sum_sig-n.x,n.y,sum_all-n.y)) %>% 
        mutate(is_significative = ifelse(p <0.01, TRUE,FALSE),
           Freq = ifelse(is.infinite(Freq),0,Freq),
           fill = ifelse(is_significative,color,"white"))

png(file.path(plotdir,paste0("enrichment_HQ_enriched.png")),width = 1000,height = 1000,res=300)
p1 = tab %>% ggplot() + geom_bar(aes(x=peak_affectation, y=Freq),fill= tab$fill, color = tab$color,
                        stat="identity") + ggtitle(paste0(i," enriched H3K27me3 peaks")) +
    theme_classic() + ggtitle(paste0("Enriched H3K27me3 peaks - ",length(which(sig))," peaks")) + 
     ylab("Log2 Enrichment") + xlab("")  + geom_hline(yintercept = 0, color ="grey")
print(p1 +  theme(axis.text.x = element_text(angle = 90, vjust =1), legend.position = "None"))
dev.off()

sig = l[["Under"]]
sum_sig = sum(table(res$peak_affectation[sig]))
sum_all = sum(table(res$peak_affectation))

tab = res[sig,] %>% count(peak_affectation, .drop=F) %>% left_join(res %>% count(peak_affectation,.drop=F), by = "peak_affectation") %>% mutate(sum_sig =sum_sig, sum_all =sum_all, Freq = log2( (n.x/sum_sig) / (n.y/sum_all) ),
color = col$col) %>% rowwise() %>% mutate(p=fisher(n.x,sum_sig-n.x,n.y,sum_all-n.y)) %>% 
        mutate(is_significative = ifelse(p <0.01, TRUE,FALSE),
           Freq = ifelse(is.infinite(Freq),0,Freq),
           fill = ifelse(is_significative,color,"white"))

png(file.path(plotdir,paste0("enrichment_HQ_under.png")),width = 1000,height = 1000,res=300)
p2= tab %>% ggplot() + geom_bar(aes(x=peak_affectation, y=Freq),fill= tab$fill, color = tab$color,
                        stat="identity") + ggtitle(paste0(i," depleted H3K27me3 peaks")) +
    theme_classic() + ggtitle(paste0("Depleted H3K27me3 peaks - ",length(which(sig))," peaks")) +
    ylab("Log2 Enrichment") + xlab("")  + geom_hline(yintercept = 0, color ="grey")
print(p2 + theme(axis.text.x = element_text(angle = 90, vjust =1), legend.position = "None"))
dev.off()


png(file.path(plotdir,paste0("enrichment_HQ_over_wolabel.png")),width = 800,height = 1000,res=300)
print(p1 + theme(text=element_blank()))
dev.off()

png(file.path(plotdir,paste0("enrichment_HQ_under_wolabel.png")),width = 800,height = 1000,res=300)
print(p2 + theme(text=element_blank()))
dev.off()

```
```{r,include=FALSE}
par(mar=c(5,5,0,0))
``` 

##5. Enrichment analysis

```{r, echo=FALSE}
###Enrichment analysis

data("hg38.GeneTSS")
GencodeGene = hg38.GeneTSS 
possibleGenes <- c(as.character(unique(GencodeGene$gene)),"ELFN2")

resTSS <- res [res$peak_affectation=="tss_pc",]

reflist <- possibleGenes
annotbase<- "MSigDB" 
MSIG.gs = hg38.MSIG.gs
enrichON <- MSIG.gs$Gene.Set[MSIG.gs$Class %in% c("c2_curated","c6_oncogenic","hallmark")] #remove unwanted gene lists to gain statistical power
MSIG.ls = hg38.MSIG.ls
database <- MSIG.ls[enrichON]

for(i in 1:length(groups)) 	{
    if (length(refs)>1) {ref <- refs[i]} else {ref <- refs[1]}
    print(paste0("Processing ",groups[i], " vs ", ref, " _ ",annotbase ))

      	signific <- which(resTSS[,paste("qval",groups[i],sep=".")] <= qval.th)
	      over     <- which(resTSS[,paste("qval",groups[i],sep=".")] <= qval.th & resTSS[,paste("log2FC",groups[i],sep=".")] > logFC.th)
	      under    <- which(resTSS[,paste("qval",groups[i],sep=".")] <= qval.th & resTSS[,paste("log2FC",groups[i],sep=".")] < (-logFC.th))
   
      if(length(over)){
        enrich.test <- geco.enrichmentTest(gene.sets=database,
                                           mylist=unique(unlist(strsplit(resTSS$gene_affectation[over],split=","))),
                                           possibleIds=reflist)
        enrich.test <- data.frame(Gene_set_name=rownames(enrich.test), enrich.test, check.names=FALSE)
        enrich.test <- merge( subset(MSIG.gs, select=-Genes), enrich.test, by.x="Gene.Set", by.y="Gene_set_name", all.y=TRUE, sort=FALSE ) ## Get class of gene set
        enrich.test <- enrich.test[order(enrich.test$`p-value`),]
        ind <- which(enrich.test$`q-value`<= qval.th);if(!length(ind)){ind <- 1:20}
        Overexpressed  <- enrich.test[ind,]	
         png(file.path(EnrichDir,paste0(groups[i],"_vs_",ref,"_qval",qval.th,"_logFC",logFC.th,"_enriched_peaks_enrichment.png")), height=1000,width=1000,res=150)
        par(mar = c(13, 4, 2, 2) + 1)
        barplot(-log10(enrich.test$`q-value`[1:10]),las=2,names.arg = enrich.test$Gene.Set[1:10],cex.names = 0.5,ylab="-log10 q-value",main=paste0("Top 10 gene list of ",length(ind)," lists"))
        dev.off()
        }
    
      if(length(under)){
        enrich.test <- geco.enrichmentTest(gene.sets=database,mylist=unique(unlist(strsplit(resTSS$gene_affectation[under],split=","))),possibleIds=reflist)
        enrich.test <- data.frame(Gene_set_name=rownames(enrich.test), enrich.test, check.names=FALSE)
        enrich.test <- merge( subset(MSIG.gs, select=-Genes), enrich.test, by.x="Gene.Set", by.y="Gene_set_name", all.y=TRUE, sort=FALSE ) ## Get class of gene set
        enrich.test <- enrich.test[order(enrich.test$`p-value`),]
        ind <- which(enrich.test$`q-value`<= qval.th);if(!length(ind)){ind <- 1:20}
        Underexpressed <- enrich.test[ind,]	
        
        png(file.path(EnrichDir,paste0(groups[i],"_vs_",ref,"_qval",qval.th,"_logFC",logFC.th,"_depleted_peaks_enrichment.png")), height=1000,width=1000,res=150)
        par(mar = c(13, 4, 2, 2) + 1)
        barplot(-log10(enrich.test$`q-value`[1:10]),las=2,names.arg = enrich.test$Gene.Set[1:10],cex.names = 0.5,ylab="-log10 q-value",main=paste0("Top 10 gene list of ",length(ind)," lists") )
        dev.off()		}
    
       WriteXLS(c("Overexpressed", "Underexpressed"), ExcelFileName =   file.path(EnrichDir,paste0("Enrichment_test_",groups[i],"_vs_",ref,"_",annotbase,"_",qval.th,"_",logFC.th,".xlsx")), SheetNames = c( paste0("Enriched_in_", groups[i]), paste0("Depleted_in_", groups[i])), perl = "perl", verbose = FALSE, row.names = FALSE, col.names = TRUE, AdjWidth = TRUE, AutoFilter = TRUE, BoldHeaderRow = TRUE, na = "", FreezeRow = 1, FreezeCol = 1)
}



```


```{r , out.width="30%",include=FALSE}
load(file.path(RDatadir,"pca.RData")) #if starting from prior analysis
load(file.path(RDatadir,"annot.RData"))#if starting from prior analysis

mati = t(pca_object)

corres_cluster <- data.frame(cluster=c("C1","C2","C3","C4","C5"),color=c("#438a5e","#6a8caf","#ff9c71","#e2979c","#b9cced"))

cor_mat <- cor(mati)
hc_PCA <- hclust(as.dist(1 - cor_mat), method = "ward.D"); gc()

intra_corr = data.frame()
for(i in unique(annot_sel2$louvain_partition)){
    cells = annot_sel2$cell_id[which(annot_sel2$louvain_partition==i)]
    tmp = cor_mat[cells,cells]
    tab = data.frame("cluster" = rep(i,ncol(tmp)), "intra_corr" = colMeans(tmp))
    intra_corr=rbind(intra_corr,tab)
}

mat.so.cor <- mati[,hc_PCA$order]
taille_groupe <- as.numeric(table(annot_sel2$louvain_partition))
IntraCorr_test <- data.frame(expressionGroup=corres_cluster$cluster,pvalue=0)


intra_corr <- intra_corr %>% dplyr::mutate(cluster=factor(cluster,levels=c("C5","C2","C1","C4","C3")))
samp = data.frame()
for(i in paste0("C",1:5)) {
  samp = rbind(samp,intra_corr[sample(which(intra_corr$cluster==i),500),])
}

#ref sample for wilcox testing, C5 cluster
MAT <- samp[samp$cluster=="C5",]
MAT_ref <- as.matrix(MAT[,-1])

for(i in 1:length(unique(annot_sel2$louvain_partition))){
  
  Group <- as.character(corres_cluster$cluster[i])
  Col_clust <- as.character(corres_cluster$color[i])
  MAT <- samp[samp$cluster==Group,]
  test <- wilcox.test(as.matrix(MAT[,-1]),MAT_ref)
  IntraCorr_test$pvalue[i] <- test$p.value

  png(file.path(resSUBdir,paste0("Pearson_Correlation_Inferno_scores_",Group,".png")), height=1150,width=1000,res=300)
  MAT <- mat.so.cor[,colnames(mat.so.cor) %in% annot_sel2$cell_id[annot_sel2$louvain_partition==Group]]
  debut <- round(100-(1-min(cor(MAT)))/0.02,0)
  image(cor(MAT),col=(inferno(100)[debut:100]))
  dev.off()
}

gc()

WriteXLS(IntraCorr_test,file.path(resSUBdir,"IntraCorr_test.xls"),row.names = T)

png(file.path(resSUBdir,"ExpressionGroup_intracorrelation.png"),width=3000,height=1500,res=300)
ggplot(dplyr::arrange(samp,cluster),aes(x=intra_corr, fill=cluster)) + geom_density(alpha=0.6) + theme_classic(base_size = 25) + 
  scale_fill_manual(values=as.character(corres_cluster$color[c(5,2,1,4,3)]))
dev.off()

png(file.path(resSUBdir,paste0("ExpressionGroup_intracorrelation_violin.png")),height=1000,width=1000,res=300)
ggplot(samp,aes(x=cluster,y=intra_corr, fill=cluster)) + 
  geom_violin(alpha=0.8) + theme_classic() + scale_fill_manual(values=as.character(corres_cluster$color)[c(5,2,1,4,3)]) +
  stat_summary(fun=median, geom="point", size=2, color="black")
dev.off()

```


5. Comparison K27me3 and K4me3 signal
```{r, echo=FALSE}
bivalenceDir = file.path("K27_peaks_K27/results/Supervised/Plots/Bivalence"); if(!dir.exists(bivalenceDir)) dir.create(bivalenceDir)

#By comparing K4 and K27 log2 RPKM at K4 peaks
K4_K27 <-  read.csv(file.path(dataDir,"results_Zerone_merge_hg38_K4_annot.csv"))

#K4 from chromatin indexing (CI) experiment (~ equivalent to bulk ChIP-seq)
K4_CI <-  read.csv(file.path(dataDir,"MM468_6_K4_chromInd.csv"))
K4_CI=K4_CI[,-c(1:3)]
K4_K27 = cbind(K4_K27, K4_CI)
rownames(K4_K27) <- paste(K4_K27$Chromosome,K4_K27$Begin,K4_K27$End,sep="_")
annotK4 <- K4_K27[,c(1:3)]
annotK4$length <- annotK4$End-annotK4$Begin
K4_K27_RPKM <- (10^9*t(t((K4_K27[,-c(1:3)]+1)/annotK4$length)/colSums(K4_K27[,-c(1:3)])))

## Replicates :
K4samples = colnames(K4_K27_RPKM)[grep("K4",colnames(K4_K27_RPKM))]
K27samples = colnames(K4_K27_RPKM)[grep("K27",colnames(K4_K27_RPKM))]

# Select 2 K4 DMSO samples
DMSO_K4_samp = c("MM468_1_J147_DMSO_K4","MM468BC_DMSO_D33_K4")
DMSO_K4_1 = K4_K27_RPKM[,DMSO_K4_samp[1]]
DMSO_K4_2 = K4_K27_RPKM[,DMSO_K4_samp[2]]

png(file.path(bivalenceDir,"smoothScatter_K4_replicates.png"),height=1000,width=1000,res=150)
smoothScatter(log2(DMSO_K4_1),log2(DMSO_K4_2),colramp=viridis,pch=19,xlab=paste0(DMSO_K4_samp[1],"  enrichment"),ylab=paste0(DMSO_K4_samp[1]," enrichment"))
dev.off()

enrichment_threshold_K4 = 2
png(file.path(bivalenceDir,paste0("Scatter_K4_replicates_enr.th_",enrichment_threshold_K4,".png")),height=1000,width=1000,res=150)
plot(log2(DMSO_K4_1),log2(DMSO_K4_2),col=alpha(ifelse((log2(DMSO_K4_1) > enrichment_threshold_K4) & (log2(DMSO_K4_2) > enrichment_threshold_K4),"red","black"),0.1),pch=19,xlab="K4me3replicate 1 enrichment",ylab="K4me3 replicate 2 enrichment")
abline(v=enrichment_threshold_K4, lty=3, lwd = 1.5, col="red")
abline(h=enrichment_threshold_K4, lty=3, lwd = 1.5, col="red")
dev.off()

# Select peaks enriched in both samples
enriched_DMSO_K4_peaks = intersect(names(DMSO_K4_1)[which(log2(DMSO_K4_1)>enrichment_threshold_K4)],
                                   names(DMSO_K4_2)[which(log2(DMSO_K4_2)>enrichment_threshold_K4)])
length(enriched_DMSO_K4_peaks)

# Select 2 K27 DMSO samples
DMSO_K27_samp = c("MM468_1_J131_DMSO_K27","MM468_3_J77_DMSO_K27")
DMSO_K27_1 = K4_K27_RPKM[,DMSO_K27_samp[1]]
DMSO_K27_2 = K4_K27_RPKM[,DMSO_K27_samp[2]]

png(file.path(bivalenceDir,"smoothScatter_K27_replicates.png"),height=1000,width=1000,res=150)
smoothScatter(log2(DMSO_K27_1),log2(DMSO_K27_2),colramp=viridis,pch=19,xlab=paste0(DMSO_K27_samp[1],"  enrichment"),ylab=paste0(DMSO_K27_samp[1]," enrichment"))
dev.off()

enrichment_threshold_K27 = 4
png(file.path(bivalenceDir,paste0("Scatter_K27_replicates_enr.th_",enrichment_threshold_K27,".png")),height=1000,width=1000,res=150)
plot(log2(DMSO_K27_1),log2(DMSO_K27_2),col=alpha(ifelse((log2(DMSO_K27_1) > enrichment_threshold_K27) & (log2(DMSO_K27_2) > enrichment_threshold_K27),"red","black"),0.1),pch=19,xlab="K27me3replicate 1 enrichment",ylab="K27me3 replicate 2 enrichment")
abline(v=enrichment_threshold_K27, lty=3, lwd = 1.5, col="red")
abline(h=enrichment_threshold_K27, lty=3, lwd = 1.5, col="red")
dev.off()

enriched_DMSO_K27_peaks = intersect(names(DMSO_K27_1)[which(log2(DMSO_K27_1)>enrichment_threshold_K27)],
                                   names(DMSO_K27_2)[which(log2(DMSO_K27_2)>enrichment_threshold_K27)])
length(enriched_DMSO_K27_peaks)

both_enriched = intersect(enriched_DMSO_K27_peaks,enriched_DMSO_K4_peaks)
length(both_enriched)

tab = data.frame(DMSO_K27_1 =log2(DMSO_K27_1), DMSO_K27_2= log2(DMSO_K27_2),
           peaks = names(DMSO_K27_1), enrichedK4 = names(DMSO_K27_1) %in% both_enriched)

png(file.path(bivalenceDir,paste0("Scatter_K27_replicates_enriched_K27_K4.png")),height=1000,width=1200,res=300)
tab %>% ggplot() + geom_point(aes(x=DMSO_K27_1,y=DMSO_K27_2,color=enrichedK4),alpha=0.3,size = 0.3) + 
    theme_classic() + scale_color_manual(values=c("grey","#47BA37"))
dev.off()

#Diff
down_res = res[which(res$log2FC.X5FU2_3_5 < -logFC.th & res$qval.X5FU2_3_5 < qval.th), ]

contingency_table_diff = data.frame( downPeaks = table(down_res$bivalent)[2:1],
                                Total= table(res$bivalent)[2:1],
                                row.names = c("Bivalent","Non bivalent"))[,c(2,4)]
contingency_table_diff

(fdiff = fisher.test(contingency_table_diff))
 
vec = c(as.numeric(contingency_table_diff[1,]/(contingency_table_diff[1,]+contingency_table_diff[2,])))
df = data.frame(percentage_bivalent = round(100*vec[2:1]),type = factor(c("All_peaks","Peaks_depleted_of_K27"),levels = c("All_peaks","Peaks_depleted_of_K27")))

library(ggpubr)
stat.test = data.frame(.y. = "percentage_bivalent",
                       group1 = "All_peaks" ,
                       group2 = "Peaks_depleted_of_K27",
                       p = paste0( "p =",formatC(fdiff$p.value, format = "e", digits = 2)),
                       p.signif = "****",
                       method="Fisher Test")

png(file.path(bivalenceDir,"Bivalent_enrichment_in_depleted_K27_peaks.png"), res=300, width = 1500, height = 1500)
df %>% ggplot(aes(x=type,y=percentage_bivalent)) + geom_bar(aes(fill=type),stat ="identity", width = 0.5) + stat_pvalue_manual(data=stat.test,label = "p", y.position = 53) + ggtitle(paste0("Bivalent peaks enrichement in H3K27me3 depleted peaks (5FU vs DMSO)")) +
    theme_classic() + scale_fill_manual(values = c("black", "forest green"))  +
    theme(axis.text.x = element_text(angle = 90, vjust =1), legend.position = "None") +
    ylab("Proportion of bivalent peaks (%)") + xlab("") + ylim(c(0,75))
dev.off()

# Down
under_res = res[which(res$log2FC.X5FU2_3_5 < -logFC.th & res$qval.X5FU2_3_5 < qval.th ), ]

contingency_table_dn = data.frame( downPeaks = table(under_res$bivalent)[2:1],
                                Total= table(res$bivalent)[2:1],
                                row.names = c("Bivalent","Non bivalent"))[,c(2,4)]
contingency_table_dn

(fdown = fisher.test(contingency_table_dn))

# Up
over_res = res[which(res$log2FC.X5FU2_3_5> logFC.th & res$qval.X5FU2_3_5 < qval.th ), ]

contingency_table_up = data.frame( upPeaks = table(over_res$bivalent)[2:1],
                                upPeaks= table(res$bivalent)[2:1],
                                row.names = c("Bivalent","Non bivalent"))[,c(2,4)]
contingency_table_up

(fup = fisher.test(contingency_table_up))

contingency_table_up_vs_down = cbind(contingency_table_up[,1],contingency_table_dn[,1])
contingency_table_up_vs_down

(fup_vs_down = fisher.test(contingency_table_up_vs_down))


vec = c(as.numeric(contingency_table_up[1,]/(contingency_table_up[1,]+contingency_table_up[2,])),
        as.numeric(contingency_table_dn[1,]/(contingency_table_dn[1,]+contingency_table_dn[2,]))[1])

df = data.frame(percentage_bivalent = round(100*vec[c(2,1,3)],2),type = factor(c("Total","Overexpressed","Underexpressed"),levels = c("Total","Overexpressed","Underexpressed")))

library(ggpubr)
stat.test = data.frame(.y. = c("percentage_bivalent","percentage_bivalent","percentage_bivalent"),
                       group1 = c("Total","Total","Overexpressed"),
                       group2 = c("Overexpressed","Underexpressed","Underexpressed"),
                       p = c(fup$p.value,fdown$p.value,fup_vs_down$p.value),
                       method=c("Fisher Test","Fisher Test","Fisher Test")
                                )
stat.test$p[which(stat.test$p > 0.05)] =  "ns"
stat.test$p[which(stat.test$p != "ns")] = paste0("p = ",formatC(as.numeric(stat.test$p[which(stat.test$p != "ns")]), format = "e", digits = 2))
png(file.path(bivalenceDir,"Bivalent_enrichment_in_diff_regions.png"), res=300, width = 1500, height = 1500)
df %>% ggplot(aes(x=type,y=percentage_bivalent)) + geom_bar(aes(fill=type),stat ="identity", width = 0.5) + stat_pvalue_manual(data=stat.test,label = "p", y.position = c(50,66,58)) + ggtitle(paste0("Bivalent peaks enriched in H3K27me3 differential peaks (5FU vs DMSO)")) +
    theme_classic() + scale_fill_manual(values = c("#000000ff","#ff0000ff","#228b22ff"))  +
    theme(axis.text.x = element_text(angle = 90, vjust =1), legend.position = "None") +
    ylab("Proportion of bivalent peaks (%)") + xlab("") + ylim(c(0,75))
dev.off()

png(file.path(bivalenceDir,"Bivalent_enrichment_in_diff_regions_HQ.png"), res=300, width = 1500, height = 1500)
stat.test.  =stat.test
stat.test.$p = c("","","")
df %>% ggplot(aes(x=type,y=percentage_bivalent)) + geom_bar(aes(fill=type),stat ="identity", width = 0.5) + stat_pvalue_manual(data=stat.test.,label = "p", y.position = c(50,66,58)) + ggtitle(paste0("Bivalent peaks enriched in H3K27me3 differential peaks (5FU vs DMSO)")) +
    theme_classic() + scale_fill_manual(values = c("#000000ff","#ff0000ff","#228b22ff"))  +
    theme(text = element_blank(), legend.position = "None") + xlab("") + ylim(c(0,75))
dev.off()
```



