---
title: "ChIPseq_unsupervised_supervised"
author: "CÃ©line Vallot"
date: "05/06/2019"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
library(here)
maindir = here()
source(file.path(maindir,"Scripts","global_var.R"))
```

```{r, echo=FALSE}
#Directories
dataDir = file.path(maindir,"input","bulk_ChIPseq")
annotDir <- file.path(maindir,"annotation")
resdir = file.path(maindir,"output","bulk_ChIPseq","MM468","K27_transcripts_10k")

#Count matrices
#result50k <- read.csv(file.path(dataDir,"results_hg38.50k.csv"))
resultZ <- read.csv(unzip(file.path(dataDir,"results_MM468_all_transcripts10k_v2.zip"),
                            exdir = tempdir()))
IDs = paste(resultZ$Chromosome,resultZ$Begin,resultZ$End,sep="_")
dups = which(duplicated(IDs))
if(length(dups)>0) {resultZ = resultZ[-dups,]}
rownames(resultZ) <- paste(resultZ$Chromosome,resultZ$Begin,resultZ$End,sep="_")

#Annot peaks
annotZerone <- read.table(unzip(file.path(annotDir,"gencode.v34.annotation.transcriptTSS_10k.bed.zip"),
                          files = "gencode.v34.annotation.transcriptTSS_10k.bed",
                          exdir = tempdir()), sep="\t")
colnames(annotZerone) <- c("chr","start","end","transcripts","Gene","strand")
annotZerone$ID = paste0(annotZerone$chr,"_",annotZerone$start,"_",annotZerone$end)
if(length(dups)>0){annotZerone = annotZerone[-dups,]}
rownames(annotZerone) <- annotZerone$ID
annotZerone$lengthPeak <- annotZerone$end - annotZerone$start

#Sample annotation
annotSample <- read.csv(file.path(dataDir,"MM468","metadata_K27.txt"),sep=" ",header = F)
colnames(annotSample) = c("Treatment","Exp")
rownames(annotSample) = gsub("_K27","",annotSample$Treatment)
```


```{r, echo=FALSE}
annotZerone <- annotZerone[rownames(resultZ),]
```

```{r, echo=FALSE}
#Preparing raw count and normalized count matrices
rownames(resultZ) <- paste(resultZ$Chromosome,resultZ$Begin,resultZ$End,sep="_")
RZ <- (resultZ[,-c(1:3,grep(pattern = "input",colnames(resultZ)))]) #RZ matrix is the raw count matrix

#create RPKM matrix by normalizing per library size and peak length
RZ_RPKM <- (10^9*t(t((RZ+1)/annotZerone$lengthPeak)/colSums(RZ)))
```


```{r, echo=FALSE}
#Selecting samples for unsupervised analysis
RZ_sel <- RZ
RZ_sel <- (RZ_sel[,-grep(pattern = "GSK",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
RZ_sel <- (RZ_sel[,-grep(pattern = "MM468_1",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis
RZ_sel <- (RZ_sel[,-grep(pattern = "spike",colnames(RZ_sel))]);dim(RZ_sel) #remove C1 clones from analysis

dim(RZ_sel)
colnames(RZ_sel)
ExpName <- "MM468_K27_transcripts_10k_bulk"

annotZerone_sel <- annotZerone; dim(annotZerone_sel)

exp.sd <- apply(RZ_sel, 1, sd)
sel2 <- order(exp.sd, decreasing = TRUE)[1:1000]
RZ_sel2 <- RZ_sel[sel2,]; dim(RZ_sel2)
annotZerone_sel2 <- annotZerone_sel[sel2,]

mat <- log2(RZ_RPKM[rownames(RZ_sel2),colnames(RZ_sel2)])
dim(mat)
hist(mat,breaks=100,main="Distribution of log2 pseudo IP")

```

```{r, echo=FALSE}
resSUBdir <- file.path(resdir, "Correlation") ; if(!file.exists(resSUBdir)){dir.create(resSUBdir)}
RDatadir <- file.path(resSUBdir,"RData") ; if(!file.exists(RDatadir)){dir.create(RDatadir)}
annotSample = annot_sel <- data.frame(
    "Name" = colnames(mat),
    "Treatment"= gsub(".*_","",
                      gsub("J...|J..","",
                           gsub("_K27","",colnames(mat)))),
    "Exp" =paste0("Exp_",gsub("_.*","",
                              gsub("MM468_","",
                                   gsub("_K27","",colnames(mat))))
    ),
    "day" = gsub("_.*","",
                 gsub(".*_J","",colnames(mat))) 
)

anocol <- geco.annotToCol2(annotS=annot_sel,annotT=annot_sel,plotLegend=T,plotLegendFile=file.path(resSUBdir,"Annotation_legends.pdf"), categCol=NULL)

control <- c("#E0E0E0","#BDBDBD","#757575","#474747")
persister_color <- c("#DCEDC8","#9CCC65","#4CAF50","#009688")
treated_UNC_color <- c("#2196F3", "#4DD0E1" )
treated_GSKJ4_color <- c("#C2185B", "#EC407A" )
res_color <- c("#FFEB3B","#FFC107","#FF9800","#FF5722")

anocol[anocol[,1]=="mediumaquamarine",1] <- "#9CCC65"
anocol[anocol[,1]=="plum1",1] <- "#BDBDBD"
anocol[anocol[,2]=="#7FC97F",2] <- "#E7F0C3"
anocol[anocol[,2]=="#BEAED4",2] <- "#F2A6A6"
anocol[anocol[,2]=="#FDC086",2] <- "#FFEADB"

save(annotSample,file=file.path(RDatadir,"annot.RData"))
save(anocol,file=file.path(RDatadir,"anocol.RData"))
save(mat,file=file.path(RDatadir,"mat.RData"))
```


##1.PCA
```{r, echo=FALSE}  
################################################################################
## PCA
################################################################################
pca <- prcomp(t(mat),center=T,scale=F)

save(pca,file=file.path(RDatadir,"pca.RData"))


## Barplot of standard deviations of each PC
#pdf(file.path(resSUBdir,"PCA_stdev_barplot.pdf"), height=6, width=6)
variances <- summary(pca)$importance["Proportion of Variance",]
variances <- variances[1:(length(variances)-1)] # remove last PC as not informative
barplot(variances, names.arg = names(variances), cex.names = 1, ylab = "Proportion of Variance", xlab="Principal components", las=2,col="royalblue",border="darkblue")
#dev.off()

## Automatic 2D PCA plots all versus all (up to the PC number specified)
# significant.PCs <- 4 # Can be determined from standard deviations plot
significant.PCs <- ncol(pca$x)-2 ; if (significant.PCs > 6) {significant.PCs <- 6}
combinations <- as.data.frame(combn(significant.PCs,2,simplify=TRUE))
extendXlimLeft <- 5 # Extend left Xlim by x percent of min
extendXlimRight <- 15 # Extend right Xlim by x percent of max
extendYlimBottom <- 5 # Extend bottom Ylim by x percent of min
extendYlimTop <- 5 # Extend top Ylim by x percent of max
TextSize <- 0.4

i=1
png(file.path(resSUBdir,paste0(ExpName,sprintf("PCA%sVs%s_plot_2D_woText.png", combinations[1,i], combinations[2,i] ))), height=900,width=800,res=150)
j=1
xlimits <- c(-20,35)
ylimits <- c(-15,15)
plot(pca$x[,combinations[1,i]],pca$x[,combinations[2,i]],col=anocol[,j],xlab=paste(colnames(pca$x)[combinations[1,i]],100*variances[combinations[1,i]]),ylab=paste(colnames(pca$x)[combinations[2,i]],100*variances[combinations[2,i]]),cex=0.8,lwd=1.5, main=colnames(anocol)[j], pch=19, xlim = xlimits, ylim = ylimits
) 
text(pca$x[,combinations[1,i]],pca$x[,combinations[2,i]],labels=annot_sel[,"Name"],pos=4,cex=TextSize)
dev.off() 

```

##2. Correlation plots  
```{r, echo=FALSE}
  ################################################################################
  ## Cor Clustering on all data
  ################################################################################
  # hmColors <- colorRampPalette(brewer.pal(9, "Blues")[-c(6:9)])(100)
 
  # colnames(mat) <- annot_sel$Name[match(colnames(mat),annot_sel$Name)]
  hc_cor <- hclust(as.dist(1 - cor(mat)), method = "ward.D")
  mat.so <- mat[,hc_cor$order]

  
  if(chRangeHM){
    for(i in 1:nrow(mat.so)){
      mat.so[i,] <- geco.changeRange(mat.so[i,],newmin=0,newmax=1)
    }
  }
  
  png(file.path(resSUBdir,paste0(ExpName,"_Clustering_correlation_matrix.png")), height=1000,width=1000,res=150)
  geco.hclustAnnotHeatmapPlot(x=cor(mat.so),
                              hc=hc_cor,
                              hmColors=hmColors,
                              anocol=anocol[hc_cor$order,],#[,ncol(cc.col):1]
                              xpos=c(0.15,0.9,0.164,0.885),
                              ypos=c(0.1,0.5,0.5,0.6,0.62,0.95),
                              dendro.cex=0.5,
                              xlab.cex=0.5,
                              hmRowNames=TRUE,
                              hmRowNames.cex=0.01
  )
 dev.off()
 
  png(file.path(resSUBdir,paste0(ExpName,"_legend_clustering.png")), height=700,width=1000,res=150)
  par(mar = c(11, 8, 8, 8))
  lims <- seq(-1, 1, length.out = 200)
  image(matrix(lims, nc = 1), col = hmColors, axes = F, 
  xlab = paste0("min ",round(min(cor(mat.so)),2)," max ",max(cor(mat.so))))
  dev.off()
   
  #save(hc_cor,file=file.path(RDatadir,"hc_cor.RData"))

  
  geco.hclustAnnotHeatmapPlot(x=cor(mat.so),
                              hc=hc_cor,
                              hmColors=hmColors,
                              anocol=anocol[hc_cor$order,],#[,ncol(cc.col):1]
                              xpos=c(0.15,0.9,0.164,0.885),
                              ypos=c(0.1,0.5,0.5,0.6,0.62,0.95),
                              dendro.cex=0.5,
                              xlab.cex=0.5,
                              hmRowNames=TRUE,
                              hmRowNames.cex=0.01 
 )
  
 ################################################################################
  ## Corrplot package
  ################################################################################
  png(file.path(resSUBdir,paste0(ExpName,"_corrplot.png")), height=700,width=1000,res=150)
  corrplot(cor(mat),order="hclust",tl.col = "black",tl.cex=0.2,cl.cex=0.3,col=hmColors)
  dev.off()
  
```
  
##4. Diff analysis on IP signal only
```{r, echo=FALSE}
options(width=180)

resdir_supervised <- file.path(resdir, "Supervised");if(!file.exists(resdir)){dir.create(resdir)}
tabdir <- file.path(resdir_supervised,"Tables");if(!file.exists(tabdir)){dir.create(tabdir)}
plotdir <- file.path(resdir_supervised,"Plots");if(!file.exists(plotdir)){dir.create(plotdir)}
EnrichDir <- file.path(tabdir,"Gene_set_analysis");if(!file.exists(EnrichDir)){dir.create(EnrichDir)}
RDataSupdir <-  file.path(resdir_supervised,"RData");if(!file.exists(RDataSupdir)){dir.create(RDataSupdir)}

MSigDBFile1 <- file.path(maindir,"annotation","hg38.MSIG.gs.rda")
MSigDBFile2 <- file.path(maindir,"annotation","hg38.MSIG.ls.rda")
load(MSigDBFile1)
load(MSigDBFile2)
MSIG.ls = hg38.MSIG.ls
MSIG.gs = hg38.MSIG.gs

myrefs <- list(
# DMSO2= annotSample[which(annotSample$Treatment=="DMSO" & annotSample$Exp=="Exp_2"),"Name"],
# DMSO3_5= annotSample[which(annotSample$Treatment=="DMSO" & annotSample$Exp %in% c ("Exp_3","Exp_5")),"Name"]
DMSO= annot_sel[which(annot_sel$Treatment=="DMSO"),"Name"]
#DMSO= annotSample[which(annotSample$Treatment=="DMSO"),"Sample"]
)
mygps <- list(
# X5FU2= annotSample[which(annotSample$Treatment=="5FUR" & annotSample$Exp=="Exp_2"),"Name"],
# X5FU3_5= annotSample[which(annotSample$Treatment=="5FUR" & annotSample$Exp %in% c ("Exp_3","Exp_5")),"Name"]
X5FU2_3_5= annotSample[which(annotSample$Treatment=="5FUR" & annotSample$Exp %in% c ("Exp_2","Exp_3","Exp_5")),"Name"]
)

refs <- names(myrefs)
groups <- names(mygps)

rownames(annot_sel) <- annot_sel$Name

#selection of peaks with at least a log2 RPKM of 1 in one sample
sel <- which(apply(RZ_RPKM[,colnames(RZ_sel2)],1,max)>1)

annotZerone. = annotZerone[,c("chr","start","end","ID")]
res <- geco.ChIPseqCompareLimma(mat=RZ,
                                metadata = annot_sel,
                                ref=myrefs,
                                groups=mygps,
                                featureTab=annotZerone.[rownames(RZ),]
)
res = cbind(res,annotZerone[,c("lengthPeak","Gene","transcripts")])

write.table(res, file.path(tabdir,"Supervised_analysis_Limma.csv"),row.names=F,quote=F,sep=";",dec="," )
save(res, file=file.path(RDataSupdir,paste("Supervised_res_Limma_object.Rdata",sep="")))
  
par(mar=c(5.1,2.1,4.1,2.1))
summaryTab <- geco.summaryChIPseqCompareLimma(restab=res,
										ref=myrefs,
							 		 groups=mygps,
  							 		 sign.th=0.1,
							 		 plotdir=plotdir,
								     fc.th=2,
										volcano_dotcex = 1,
										volcano_dot_alpha = 0.2,
										volcano_add_xlim = 2,
										display_top_names=0
										)
									 
summaryTab <- data.frame("."=row.names(summaryTab), summaryTab, check.names=FALSE)		
write.table(summaryTab, file.path(tabdir,"Number_differentially_bound_peaks_per_group.csv"),row.names=F,quote=F,sep=";"  )

summaryTab <- geco.summaryChIPseqCompareLimma(restab=res,
										ref=myrefs,
							 		 groups=mygps,
  							 		 sign.th=0.1,
							 		 plotdir=plotdir,
								     fc.th=2,
										volcano_dotcex = 1,
										volcano_dot_alpha = 0.2,
										volcano_add_xlim = 2,
										display_top_names=10
										)
```
```{r,include=FALSE}
par(mar=c(5,5,0,0))
``` 

##5. Enrichment analysis

```{r, echo=FALSE}
###Enrichment analysis

data("hg38.GeneTSS")
GencodeGene = hg38.GeneTSS 
possibleGenes <- unique(GencodeGene$gene)


qval.th <- 0.1
for(FC in c(1,2)){
  resTSS <- res #[res$peak_affectation=="tss_pc",]
  #resTSS <- res[abs(as.numeric(res$dist_enhancer))<1000,] #you can change the annotation of interest to link peaks to genes
  reflist <- possibleGenes
  annotbase<- "MSigDB" 
  MSIG.gs = hg38.MSIG.gs
  enrichON <- MSIG.gs$Gene.Set[MSIG.gs$Class %in% c("c2_curated","c6_oncogenic","hallmark")] #remove unwanted gene lists to gain statistical power
  MSIG.ls = hg38.MSIG.ls
  database <- MSIG.ls[enrichON]
  resTSS$Gene = as.character(resTSS$Gene)
  resTSS = resTSS[which(resTSS$Gene %in% hg38.GeneTSS$gene),]
  
  for(i in 1:length(groups)) 	{
    if (length(refs)>1) {ref <- refs[i]} else {ref <- refs[1]}
    print(paste0("Processing ",groups[i], " vs ", ref, " _ ",annotbase ))
    
    signific <- which(resTSS[,paste("qval",groups[i],sep=".")] <= qval.th)
    over     <- which(resTSS[,paste("qval",groups[i],sep=".")] <= qval.th & resTSS[,paste("log2FC",groups[i],sep=".")] > FC)
    under    <- which(resTSS[,paste("qval",groups[i],sep=".")] <= qval.th & resTSS[,paste("log2FC",groups[i],sep=".")] < (-FC))
    
    if(length(over)){
      enrich.test <- geco.enrichmentTest(gene.sets=database,
                                         mylist=unique(unlist(strsplit(resTSS$Gene[over],split=","))),
                                         possibleIds=reflist)
      enrich.test <- data.frame(Gene_set_name=rownames(enrich.test), enrich.test, check.names=FALSE)
      enrich.test <- merge( subset(MSIG.gs, select=-Genes), enrich.test, by.x="Gene.Set", by.y="Gene_set_name", all.y=TRUE, sort=FALSE ) ## Get class of gene set
      enrich.test <- enrich.test[order(enrich.test$`p-value`),]
      ind <- which(enrich.test$`q-value`<= qval.th);if(!length(ind)){ind <- 1:20}
      Overexpressed  <- enrich.test[ind,]	
      png(file.path(EnrichDir,paste0(groups[i],"_vs_",ref,"_qval",qval.th,"_logFC",FC,"_enriched_peaks_enrichment.png")), height=1000,width=1000,res=150)
      par(mar = c(13, 4, 2, 2) + 1)
      barplot(-log10(enrich.test$`q-value`[1:10]),las=2,names.arg = enrich.test$Gene.Set[1:10],cex.names = 0.5,ylab="-log10 q-value",main=paste0("Top 10 gene list of ",length(ind)," lists"))
      dev.off()
    }
    
    if(length(under)){
      enrich.test <- geco.enrichmentTest(gene.sets=database,mylist=unique(unlist(strsplit(resTSS$Gene[under],split=","))),possibleIds=reflist)
      enrich.test <- data.frame(Gene_set_name=rownames(enrich.test), enrich.test, check.names=FALSE)
      enrich.test <- merge( subset(MSIG.gs, select=-Genes), enrich.test, by.x="Gene.Set", by.y="Gene_set_name", all.y=TRUE, sort=FALSE ) ## Get class of gene set
      enrich.test <- enrich.test[order(enrich.test$`p-value`),]
      ind <- which(enrich.test$`q-value`<= qval.th);if(!length(ind)){ind <- 1:20}
      Underexpressed <- enrich.test[ind,]	
      
      png(file.path(EnrichDir,paste0(groups[i],"_vs_",ref,"_qval",qval.th,"_logFC",FC,"_depleted_peaks_enrichment.png")), height=1000,width=1000,res=150)
      par(mar = c(13, 4, 2, 2) + 1)
      barplot(-log10(enrich.test$`q-value`[1:10]),las=2,names.arg = enrich.test$Gene.Set[1:10],cex.names = 0.5,ylab="-log10 q-value",main=paste0("Top 10 gene list of ",length(ind)," lists") )
      dev.off()		}
    
    WriteXLS(c("Overexpressed", "Underexpressed"), ExcelFileName =   file.path(EnrichDir,paste0("Enrichment_test_",groups[i],"_vs_",ref,"_",annotbase,"_",qval.th,"_",FC,".xlsx")), SheetNames = c( paste0("Enriched_in_", groups[i]), paste0("Depleted_in_", groups[i])), perl = "perl", verbose = FALSE, row.names = FALSE, col.names = TRUE, AdjWidth = TRUE, AutoFilter = TRUE, BoldHeaderRow = TRUE, na = "", FreezeRow = 1, FreezeCol = 1)
  }
}
```




