---
    title: "PDX AllBC1224 scRNAseq"
author: CÃ©line Vallot
date: 21/04/2020
output: html_document
---
    
    ```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE)
```

```{r libraries}
library(here)
source(file.path(here(),"Scripts","global_var.R"))

```


```{r paths}
#Paths
resdir <- file.path(here(),"output","scRNAseq","PDX_BC1224")
QCdir <- file.path(resdir,"QC")
resSUBdir <- file.path(resdir, "Unsupervised") ; if(!file.exists(resSUBdir)){dir.create(resSUBdir)}
resSUBdir_UMAP <- file.path(resdir, "Unsupervised","UMAP") ; if(!file.exists(resSUBdir_UMAP)){dir.create(resSUBdir_UMAP)}
resSUBdir_boxplots <- file.path(resdir, "Unsupervised","boxplots") ; if(!file.exists(resSUBdir_boxplots)){dir.create(resSUBdir_boxplots)}
resSUBdir_heatmaps = file.path(resSUBdir,"Heatmaps"); if(!dir.exists(resSUBdir_heatmaps)) dir.create(resSUBdir_heatmaps)

resSUBdir_sub = file.path(resSUBdir,"sub500"); if(!dir.exists(resSUBdir_sub)) dir.create(resSUBdir_sub)
RDatadir <- file.path(resSUBdir,"RData") ; if(!file.exists(RDatadir)){dir.create(RDatadir)}

resdir_BC1224 <- file.path(here(),"output","scRNAseq","PDX")
RDatadir_BC1224 = file.path(resdir_BC1224, "Unsupervised", "RData")
```

```{r loads}
load(file.path(RDatadir,"BC1224.RData"))
annotationDatabases <- c("MSigDB") # "KEGG" or "GO" or "KEGG_GO" or "MSigDB"

cells = annot$cell_id[which(annot$total_features > 1500)]
annot = annot[match(cells, annot$cell_id),]
Signal = Signal[,match(cells, colnames(Signal))]
```

We can start from already performed analysis to make figures.

### Analyzing all cells from v3 10X kits with all genes

Here, we use Monocle3 to reduce dimensions using PCA and UMAP, based on all genes. Cells are colored according to sample_id, cluster membership (louvain), expression of ribosomal genes and total counts.

### Studying cell cycle with Seurat, calculate a cell cycle score

Here we use CellCycleScoring function to annotate cells with a cell cycle phase.
By the way we perform as a comparison PCA and UMAP reduction using Seurat. The main difference with our prior analysis is that Seurat performs PCA solely on the most varying genes (n=2000) and not the entire dataset. The segragation of cells seems less efficient in 2-D space then with all genes.

```{r, echo=FALSE}
library(monocle3)
# A list of cell cycle markers, from Tirosh et al, 2015, is loaded with Seurat.  We can
# segregate this list into markers of G2/M phase and markers of S phase
s.genes <- cc.genes$s.genes
g2m.genes <- cc.genes$g2m.genes

# Bind Signal & annots 
# genes_in_common = intersect(rownames(BC1224$Signal), rownames(Signal))
# Signal = Signal[match(genes_in_common,rownames(Signal)),]
# BC1224$Signal = BC1224$Signal[match(genes_in_common,rownames(BC1224$Signal)),]
# Signal = cbind(BC1224$Signal, Signal)
# 
# gene_metadata = gene_metadata[match(genes_in_common,rownames(gene_metadata)),]
# annot = rbind(BC1224$annot,annot)

# Create our Seurat object and complete the initalization steps
rownames(Signal) <- gene_metadata$Symbol
annot_sel <- annot[annot$kit %in% c("v3"),]

marrow <- CreateSeuratObject(counts = Signal[,which(colnames(Signal) %in% annot_sel$cell_id)])
table(marrow$orig.ident)

# Integration
# marrow.list <- SplitObject(marrow, split.by = "model")
# perform standard preprocessing on each object
# for (i in 1:length(marrow.list)) {
#   marrow.list[[i]] <- NormalizeData(marrow.list[[i]],  verbose = FALSE)
#   marrow.list[[i]] <- FindVariableFeatures(marrow.list[[i]], selection.method = "vst")
# }

# marrow.list <- lapply(X = marrow.list, FUN = SCTransform)
# features <- SelectIntegrationFeatures(marrow.list)

# marrow.list <- PrepSCTIntegration(
#   marrow.list,
#   anchor.features = features
# )
# 
# anchors <- FindIntegrationAnchors(
#   marrow.list,
#   normalization.method = "SCT",
#   anchor.features = features
# )

# marrow <- IntegrateData(anchors)
# DefaultAssay(marrow) <- "RNA"

marrow <- NormalizeData(marrow,  verbose = FALSE)
marrow <- FindVariableFeatures(marrow, selection.method = "vst")
marrow <- ScaleData(marrow, features = rownames(marrow))
marrow <- RunPCA(marrow, features = VariableFeatures(marrow), ndims.print = 1:10, nfeatures.print = 10)
DimPlot(marrow, reduction = "pca")
FeaturePlot(marrow, reduction = "pca", features = "nFeature_RNA", cols = rev(viridis::magma(100)))
DimHeatmap(marrow, dims = 1:3, cells = 1000, balanced = TRUE)
ElbowPlot(marrow)
marrow <- FindNeighbors(marrow, dims = 1:20)
marrow <- RunUMAP(marrow, dims = 1:20)
marrow <- Seurat::FindClusters(marrow)

marrow$orig.ident = sub("_[A-Z]*$","",colnames(marrow))

DimPlot(marrow, group.by = "orig.ident",reduction = "umap")
FeaturePlot(marrow,  features = "nCount_RNA",reduction = "umap", cols = rev(viridis::magma(100)))
FeaturePlot(marrow,  features = "nFeature_RNA",reduction = "umap", cols = rev(viridis::magma(100)))

marrow <- CellCycleScoring(marrow, s.features = s.genes, g2m.features = g2m.genes, set.ident = TRUE)
annot_seurat <- data.frame (Seurat_cluster=paste0("C",marrow$seurat_clusters),cell_cycle=marrow$Phase)
annot_seurat$cell_id <- rownames(annot_seurat)
if(all.equal(rownames(annot_sel),rownames(annot_seurat))) annot_int <- cbind(annot_sel,annot_seurat) else print("ERROR")

save(marrow, file = file.path(RDatadir,"SeuratObject.RData"))
# rm(marrow); gc()

# load(file = file.path(RDatadir,"SeuratObject.RData"))
# marrow_BC408 = marrow
# FeaturePlot(marrow_BC408, features = c("KDM6B","IGFBP3","KRT6A","KRT14"))
# FeaturePlot(marrow_BC408, features =c("PPARG","FOSL1","THAP1","THAP11"))
# FeaturePlot(marrow_BC408, features =c("NFKB1","PBX1","EGR4","SMARCC1"))
# FeaturePlot(marrow_BC408, features =c("FOSB","JUNB","JUND","JUN"))
# FeaturePlot(marrow_BC408, features =c("FOSL1","FOSL2"), pt.size = 0.2)
# FeaturePlot(marrow_BC408, features =c("KRTDAP","KRT6A", "KRT75", "NNMT"))
# 
# FeaturePlot(marrow_BC408, features =c("NEXN"))
# FeaturePlot(marrow, features =c("NEXN"))


```

###Study cells with PCA, UMAP & hierarchical clustering

```{r, echo=FALSE}

cells_low_rRNA = ( (marrow@meta.data$nFeature_RNA / marrow@meta.data$nCount_RNA) < 0.2)
table(cells_low_rRNA)

Signal = Signal[,cells_low_rRNA]
annot_sel = annot_sel[cells_low_rRNA,]

cells_high_count = ( annot_sel$total_counts > 25000)

Signal = Signal[,cells_high_count]
annot_sel = annot_sel[cells_high_count,]

rownames(Signal) <- gene_metadata$Symbol
annot_sel <- annot_sel[,!duplicated(colnames(annot_sel))]
rownames(annot_sel) <- annot_sel$cell_id
rownames(gene_metadata) <- gene_metadata$Symbol

cds <- new_cell_data_set(Signal[,annot_sel$cell_id],
                         cell_metadata = annot_sel,
                         gene_metadata  = gene_metadata)

cds <- preprocess_cds(cds, method='PCA', norm_method='size_only',num_dim=50)
monocle3::plot_pc_variance_explained(cds)

anocol = scTools::geco.annotToCol4(as.data.frame(cds@colData))


pdf(file.path(resSUBdir_UMAP,"PC_total_features.pdf"))
for(i in 1 : 10){
    for(j in i:10){
        print(
            plot(cds@int_colData@listData$reducedDims$PCA[,c(i,j)], col = anocol[,"total_features"], pch = 20 )
        )
    }
}
dev.off()


cds <- reduce_dimension(cds, reduction_method = 'UMAP')
cds <-  cluster_cells(cds, k = 5, resolution = 0.005)

pdf(file.path(resSUBdir_UMAP,"monocle_UMAPs.pdf"))
monocle3::plot_cells(cds, color_cells_by = "sample_id",reduction_method = "UMAP", 
                     group_label_size = 4, cell_stroke =2 )
monocle3::plot_cells(cds, color_cells_by = "total_counts",reduction_method = "UMAP", 
                     group_label_size = 4, cell_stroke =2 )
monocle3::plot_cells(cds, color_cells_by = "total_features",reduction_method = "UMAP", 
                     group_label_size = 4, cell_stroke =2 )
monocle3::plot_cells(cds, color_cells_by = "doublet_score",reduction_method = "UMAP", 
                     group_label_size = 4, cell_stroke =2 )
monocle3::plot_cells(cds, color_cells_by = "cluster",reduction_method = "UMAP", 
                     group_label_size = 4, cell_stroke =2 )
dev.off()


annot_sel$louvain_cluster <- cds@clusters[[1]]$clusters
annot_sel$louvain_cluster <- paste0("C",annot_sel$louvain_cluster)

# annot_int$louvain_partition <- cds@clusters[[1]]$partition
# annot_int$louvain_partition <- paste0("C",annot_int$louvain_partition)

umap_res <- cds@int_colData$reducedDims$UMAP # with R3.6.2 and higher for SummarizedExperiment, otherwise cds@reducedDims[[2]]
pca_object <-cds@int_colData$reducedDims$PCA

save(umap_res,file=file.path(RDatadir,"umap.RData"))
save(pca_object,file=file.path(RDatadir,"pca.RData"))

NormCounts <- t(t(exprs(cds)) /  pData(cds)[, 'Size_Factor']) #keep format sparse matrix, need to have row and colnames separate
LogCounts <- log(NormCounts+1,2)

save(NormCounts,file=file.path(RDatadir,"NormCounts.RData"))
save(LogCounts,file=file.path(RDatadir,"LogCounts.RData"))
save(annot_sel,file=file.path(RDatadir,"annot_int.RData"))
rm(NormCounts); gc()
# rm(cds); gc()


```


### Perform hierarchical clustering
```{r , out.width="30%",include=FALSE}

mati <- t(pca_object)
hc_PCA <- hclust(as.dist(1 - cor(mati)), method = "ward.D")
mat.so.cor <- mati[,hc_PCA$order]

nclust <- 3
annot_sel$clustCol3 <- paste("C",match(cutree(hc_PCA,nclust),unique(cutree(hc_PCA,nclust)[hc_PCA$order])),sep="")
nclust <- 4
annot_sel$clustCol4 <- paste("C",match(cutree(hc_PCA,nclust),unique(cutree(hc_PCA,nclust)[hc_PCA$order])),sep="")
nclust <- 5
annot_sel$clustCol5 <- paste("C",match(cutree(hc_PCA,nclust),unique(cutree(hc_PCA,nclust)[hc_PCA$order])),sep="")
nclust <- 6
annot_sel$clustCol6 <- paste("C",match(cutree(hc_PCA,nclust),unique(cutree(hc_PCA,nclust)[hc_PCA$order])),sep="")
```

```{r, echo=FALSE}
pcaText <- FALSE
annotText <- "sample_id"
hcText <- "sample_id"  
load(file.path(RDatadir,"LogCounts.RData"))

annotCol <- c("sample_id","total_features","cell_cycle","rRNA", "louvain_cluster")
pcaText <- FALSE
annotText <- "sample_id"

ribo <- grepl("^RP[SL]",gene_metadata$Symbol)
annot_sel$rRNA <- apply(LogCounts[ribo,],2,mean)
annot_sel$cell_cycle <- annot_seurat$cell_cycle[match(annot_sel$cell_id,annot_seurat$cell_id)]
# annot_sel$cell_cycle = annot_seurat$cell_cycle[match(annot_int$cell_id,annot_seurat$cell_id)]
# annot_sel$Seurat_custer = annot_seurat$Seurat_cluster[match(annot_int$cell_id,annot_seurat$cell_id)]
anocol <- geco.annotToCol4(annotS=annot_sel[,annotCol],annotT=annot_sel,plotLegend=T,plotLegendFile=file.path(resSUBdir,"Annotation_legends.pdf"), scale_q = "inferno")

#Common palette for all MM468 experiments
control <- c("#E0E0E0","#BDBDBD","#757575","#ffcbcb")
persister_color <- c("#DCEDC8","#9CCC65","#4CAF50","#009688")
res_color <- c("#FFEB3B","#FFC107","#FF9800","#FF5722")
color_PDX <- c(control[c(1)],persister_color[c(3)])


samples_PDX = c("BC1224_chemonaive", "BC1224_persister")
corres <- data.frame(PDX=samples_PDX,color=color_PDX)
anocol[,"sample_id"] <- as.character(corres$color[match(annot_sel$sample_id,corres$PDX)])

corres_cell_cycle <- data.frame(phase=c("G1","G2M","S"),color=c("#e896aaff","#553fc2ff","#5d6c7dff"))
indice <- which(annotCol=="cell_cycle")
anocol[,indice] <- as.character(corres_cell_cycle$color[match(annot_sel$cell_cycle,corres_cell_cycle$phase)])

corres_cluster <- data.frame(phase=c("C1","C2","C3","C4","C5","C6"),color=c("#6a8caf","#e2979c","#b9cced","#438a5e","#FFC107","#009688"))

# indice <- which(annotCol=="clustCol5")
# anocol[,indice] <- as.character(corres_cluster$color[match(annot_int$clustCol5,corres_cluster$phase)])
indice <- which(annotCol=="louvain_cluster")
anocol[,indice]  <- as.character(corres_cluster$color[match(annot_sel$louvain_cluster,corres_cluster$phase)])

png(file.path(resSUBdir,"Legend_sample_scRNAseq.png"), height=2000,width=1500,res=300)
barplot(rep(1,2),col=corres$color, cex.names  =0.8,horiz=F,names.arg=corres$PDX,las=2)
dev.off()

png(file.path(resSUBdir,"Legend_cluster_scRNAseq.png"), height=2000,width=1500,res=300)
barplot(rep(1,6),col=corres_cluster$color, cex.names  =0.8,horiz=F,names.arg=c("C1","C2","C3","C4","C5","C6"),las=2)
dev.off()

png(file.path(resSUBdir_boxplots,"Boxplot_rRNA.png"),width=1500,height=1500,res=300)
boxplot(annot_sel$rRNA~annot_sel$sample_id,las=2)
dev.off()

test <- wilcox.test(annot_sel$rRNA[which(annot_sel$sample_id %in% c("BC1224_persister"))],
                    annot_sel$rRNA[which(annot_sel$sample_id == "BC1224_chemonaive")]) 
save(anocol,annot_int,file=file.path(RDatadir,"annot_anocol_final.RData")) 
```

###Generate figures: UMAP and clustering with correlation heatmap

```{r, out.width="30%", echo=FALSE}
load(file.path(RDatadir,"umap.RData"))


for(j in 1:ncol(anocol))
{
    png(file.path(resSUBdir_UMAP,paste0("UMAP_",colnames(anocol)[j],".png")), height=1350,width=1200,res=300)
    if(class(annot_sel[1,colnames(anocol)[j]])=="numeric"){
        plot((umap_res), col=alpha(anocol[,j],1),pch=20,cex=0.6,
             main=paste0(colnames(anocol)[j]," min=",round(min(annot_sel[,colnames(anocol)[j]]),digits=3)," max=",round(max(annot_sel[,colnames(anocol)[j]]),digits=3)),
             xlab="component 1",ylab="component 2")} else {
                 plot((umap_res), col=alpha(anocol[,j],1),pch=20,cex=0.6,
                      main=paste0(colnames(anocol)[j]),
                      xlab="component 1",ylab="component 2")
             }
    
    
    dev.off()
}

# Subsample 
umap_res_sub = umap_res
anocol_sub = anocol
annot_int_sub = annot_int
LogCounts_sub = LogCounts
gc()
save(umap_res_sub,anocol_sub,annot_int_sub,LogCounts_sub, file = file.path(RDatadir,"all_sub_500.RData"))

for(j in 1:ncol(anocol_sub))
{
    png(file.path(resSUBdir_sub,paste0("UMAP_sub500_",colnames(anocol_sub)[j],".png")), height=1350,width=1200,res=300)
    if(class(annot_int_sub[1,colnames(anocol_sub)[j]])=="numeric"){
        plot((umap_res_sub), col=alpha(anocol_sub[,j],0.7),pch=20,cex=0.6,
             main=paste0(colnames(anocol_sub)[j]," min=",round(min(annot_int_sub[,colnames(anocol_sub)[j]]),digits=3)," max=",round(max(annot_int_sub[,colnames(anocol_sub)[j]]),digits=3)),
             xlab="component 1",ylab="component 2")} else {
                 plot((umap_res_sub), col=alpha(anocol_sub[,j],0.7),pch=20,cex=0.6,
                      main=paste0(colnames(anocol_sub)[j]),
                      xlab="component 1",ylab="component 2")
             }
    
    
    dev.off()
}

png(file.path(resSUBdir,"Clustering_correlation_matrix_PCA_all_annot.png"), height=1500,width=1500,res=300)
geco.hclustAnnotHeatmapPlot(x=cor(mat.so.cor),
                            hc=hc_PCA,
                            hmColors=corColors,
                            anocol=anocol[hc_PCA$order,c("sample_id","total_features","louvain_cluster")],#[,ncol(cc.col):1]
                            xpos=c(0.15,0.9,0.164,0.885),
                            ypos=c(0.1,0.5,0.5,0.6,0.62,0.95),
                            dendro.cex=0.1,
                            xlab.cex=0.2,
                            hmRowNames=FALSE,
                            hmRowNames.cex=0.01
)
dev.off()


ClusterMembership <- as.data.frame.matrix(table(annot_sel$sample_id,annot_sel$louvain_cluster))
ClusterMembership$pvalue <- 0
ClusterMembership$sample_id <- rownames(ClusterMembership)
for(i in c(1:(dim(ClusterMembership)[1]))){
    test <- fisher.test(ClusterMembership[c(i,1),c(1:2)]) 
    ClusterMembership$pvalue[i] <- test$p.value
}

WriteXLS(ClusterMembership,file.path(resSUBdir_boxplots,"ClusterMembership.xls"),row.names = T)

```



###Study cell cycle in every sample

```{r , out.width="30%",include=FALSE}
#Association of cell cycle to samples
CellCycle <- as.data.frame.matrix(table(annot_int$sample_id,annot_int$cell_cycle))
CellCycle$pvalue <- 0
for(i in c(1:(dim(CellCycle)[1]))){
    test <- chisq.test(CellCycle[c(i,1),c(1:3)]) 
    CellCycle$pvalue[i] <- test$p.value
}

WriteXLS(CellCycle,file.path(resSUBdir_boxplots,"CellCycle_samples_vs_UNT.xls"), row.names = T)
CellCycle$sample <- rownames(CellCycle)
CellCycle_b <- CellCycle[,-4] %>% tidyr::pivot_longer(cols=c("G1","G2M","S"),names_to = "cycle",values_to = "freq") %>% dplyr::mutate(cycle=factor(cycle, levels =c("G1","G2M","S"))) %>% dplyr::group_by(sample) %>% dplyr::mutate(freq_prop=freq/sum(freq))
png(file.path(resSUBdir_boxplots,"CellCycle.png"),width=1000,height=1000,res=300)
ggplot(CellCycle_b) +geom_col(aes(y=freq_prop,x=sample,fill=cycle)) + 
    scale_fill_manual(values=as.character(corres_cell_cycle$color)) + 
    theme_classic() +theme(axis.text.x=element_text(angle=90, hjust=1)) +
    theme(text=element_blank()) # if no legend
dev.off()

```

## Supervised 

```{r , out.width="30%",include=FALSE}
resdir_supervised=file.path(resdir,"Supervised"); if(!dir.exists(resdir_supervised)) dir.create(resdir_supervised)
tabdir <- file.path(resdir_supervised,"Tables");if(!file.exists(tabdir)){dir.create(tabdir)}
plotdir <- file.path(resdir_supervised,"Plots");if(!file.exists(plotdir)){dir.create(plotdir)}
EnrichDir <- file.path(tabdir,"Gene_set_analysis");if(!file.exists(EnrichDir)){dir.create(EnrichDir)}
RDataSupdir <-  file.path(resdir_supervised,"RData");if(!file.exists(RDataSupdir)){dir.create(RDataSupdir)}
resdir_boxplots <-  file.path(resdir_supervised,"Boxplots");if(!file.exists(resdir_boxplots)){dir.create(resdir_boxplots)}
resdir_UMAPs <-  file.path(resdir_supervised,"UMAP");if(!file.exists(resdir_UMAPs)){dir.create(resdir_UMAPs)}
resdir_heatmaps <-  file.path(resdir_supervised,"Heatmaps");if(!file.exists(resdir_heatmaps)){dir.create(resdir_heatmaps)}

load(file.path(resSUBdir,"RData","BC1224.RData"))
RawCounts = Signal
gc()
metadata = annot_int
feature = gene_metadata

#Get the gene list of Persister cells (6souris) overexpressed
mygps <- list(
    'persister'=metadata[which(metadata$louvain_cluster %in% c("C4","C6") ),"cell_id"]
)
sapply(mygps,length)

myrefs <- list(
    'chemonaive'=metadata[which(!metadata$louvain_cluster %in% c("C4","C6") ),"cell_id"]
)
sapply(myrefs,length)

refs <- names(myrefs) 
groups <- names(mygps) 

algoType <- c("edgeR") ## Limma or DESeq or a vector of both or edgeR for single-cell
useBlockFactor <- FALSE ## TRUE or FALSE

my.res <- scTools::geco.CompareedgeRGLM(dataMat=as.matrix(RawCounts),
                                annot=metadata,
                                ref=myrefs,
                                groups=mygps,
                                featureTab=feature
)

log2FC_thresholds <- log2(c(3))
Signif_threshold <- 0.01

for(log2FC_threshold in log2FC_thresholds){
    for(i in 1:length(myrefs)){
        ref = names(myrefs)[i]
        logFC_name = (paste0("log2FC.persister"))
        qval_name = (paste0("qval.persister"))
        count_name = (paste0("logCPM.persister"))
        cat("Saving for FC =",2^log2FC_threshold,"\n")
        under_res = my.res %>% dplyr::filter(
            .data[[logFC_name]] < -log2FC_threshold &
                .data[[qval_name]] < Signif_threshold) %>% 
            dplyr::arrange(.data[[qval_name]]) %>%
            dplyr::select(Symbol,.data[[logFC_name]],.data[[qval_name]])
        
        over_res = my.res %>% dplyr::filter(
            .data[[logFC_name]] > log2FC_threshold &
                .data[[qval_name]] < Signif_threshold) %>% 
            dplyr::arrange(.data[[qval_name]]) %>%
            dplyr::select(Symbol,.data[[logFC_name]],.data[[qval_name]])
        
        res_select = my.res[,c("Symbol",logFC_name,qval_name,count_name)]
        WriteXLS(c("over_res","under_res","my.res"),
                 ExcelFileName = file.path(
                     tabdir,paste0("Differential_analysis_Limma_logFC_Pers_vs_",ref,"_",
                                   round(log2FC_threshold,2),".xlsx")),
                 SheetNames = c(
                     paste0("Over_",round(log2FC_threshold,2),"_",Signif_threshold,"_n",
                            nrow(over_res)),paste0(
                                "Under_-",round(log2FC_threshold,2),"_",Signif_threshold,
                                "_n",nrow(under_res)),"All"),
                 perl = "perl", verbose = FALSE, row.names = FALSE,
                 col.names = TRUE, AdjWidth = T, AutoFilter = TRUE,
                 BoldHeaderRow = TRUE, na = "", FreezeRow = 1, FreezeCol = 1)
        
        summaryTab <- scTools::geco.summaryCompareedgeR(restab=my.res,
                                               ref=myrefs,
                                               groups=mygps,
                                               qval.th=Signif_threshold,
                                               fc.th=log2FC_threshold,
                                               plotdir=plotdir
        )
        
        summaryTab <- data.frame("."=row.names(summaryTab), summaryTab,
                                 check.names=FALSE)		
        write.table(summaryTab, file.path(
            tabdir,paste0("Number_differentially_expressed_genes_per_group_logFC_",
                          round(log2FC_threshold,2),".csv")),row.names=F,quote=F,sep=";")
    }
}
save(my.res, file=file.path(RDataSupdir,paste("Supervised_res_object_edgeR.Rdata",sep="")))
print("summaryedgeRCompare Done")

```

### Run Gene Set Analysis

```{r , out.width="30%",include=FALSE}
#import my.res et gene lists from diff analysis, logCounts, take top 10
MSigDBFile1 <- "../../../annotation/hg38.MSIG.gs.rda" 
MSigDBFile2 <- "../../../annotation/hg38.MSIG.ls.rda"
load(file=file.path(RDataSupdir,paste("Supervised_res_object_edgeR.Rdata",sep="")))
load(MSigDBFile1)
load(MSigDBFile2)
MSIG.ls = hg38.MSIG.ls
MSIG.gs = hg38.MSIG.gs

rownames(my.res) <- my.res$Gene

annotbase <- "MSigDB" 
database <- MSIG.ls ##MSigDB

Overexpressed  <- Underexpressed <- data.frame()

reflist <- unique(my.res$Symbol);length(reflist)

for(i in 1:length(groups)) 	{
    for(log2FC_threshold in log2FC_thresholds){
        gp <- groups[i]
        if (length(refs)>1) {ref <- refs[i]} else {ref <- refs[1]}
        print(paste0("Processing ",gp, " vs ", ref, " _ ",annotbase," for logFC = ",round(log2FC_threshold,2)))
        
        signific <- which(my.res[,paste("qval",gp,sep=".")] <= Signif_threshold & abs(my.res[,paste("log2FC",gp,sep=".")]) > log2FC_threshold)
        over <- which(my.res[,paste("qval",gp,sep=".")] <= Signif_threshold & my.res[,paste("log2FC",gp,sep=".")] > log2FC_threshold)
        under <- which(my.res[,paste("qval",gp,sep=".")] <= Signif_threshold & my.res[,paste("log2FC",gp,sep=".")] < -log2FC_threshold)
        print(paste0("significant = ", length(signific))) ; print(paste0("over = ", length(over))) ; print(paste0("under = ", length(under)))
        
        
        if(length(over)){
            enrich.test <- geco.enrichmentTest(gene.sets=database,mylist=my.res$Symbol[over],possibleIds=reflist)
            enrich.test <- data.frame(Gene_set_name=rownames(enrich.test), enrich.test, check.names=FALSE)
            enrich.test <- merge( subset(MSIG.gs, select=-Genes), enrich.test, by.x="Gene.Set", by.y="Gene_set_name", all.y=TRUE, sort=FALSE ) ## Get class of gene set
            
            enrich.test <- enrich.test[order(enrich.test$`p-value`),]
            # ind <- which(enrich.test$`q-value`<= 0.1);if(!length(ind)){ind <- 1:20}
            Overexpressed  <- enrich.test #[ind,]
            png(file.path(EnrichDir,paste0(groups[i],"_vs_",ref,"_qval",Signif_threshold,"_logFC",round(log2FC_threshold,2),"_enriched_peaks_enrichment.png")), height=1000,width=1000,res=150)
            par(mar = c(13, 4, 2, 2) + 1)
            barplot(-log10(enrich.test$`q-value`[1:10]),las=2,names.arg = enrich.test$Gene.Set[1:10],cex.names = 0.5,ylab="-log10 q-value",main=paste0("Top 10 gene list of ",10," lists"))
            dev.off()
        }
        
        if(length(under)){
            enrich.test <- geco.enrichmentTest(gene.sets=database,mylist=my.res$Symbol[under],possibleIds=reflist)
            enrich.test <- data.frame(Gene_set_name=rownames(enrich.test), enrich.test, check.names=FALSE)
            enrich.test <- merge( subset(MSIG.gs, select=-Genes), enrich.test, by.x="Gene.Set", by.y="Gene_set_name", all.y=TRUE, sort=FALSE ) ## Get class of gene set
            
            enrich.test <- enrich.test[order(enrich.test$`p-value`),]
            # ind <- which(enrich.test$`q-value`<= 0.1);if(!length(ind)){ind <- 1:20}
            Underexpressed <- enrich.test #[ind,]
            }
        
        WriteXLS(c("Overexpressed", "Underexpressed"), ExcelFileName = file.path(EnrichDir,paste0("Enrichment_test_",gp,"_vs_",ref,"_",annotbase,"_logFC",round(log2FC_threshold,2),"_unfiltered.xlsx")), SheetNames = c( paste0("Overexp_in_", gp), paste0("Underexp_in_", gp) ), perl = "perl", verbose = FALSE, row.names = FALSE, col.names = TRUE, AdjWidth = TRUE, AutoFilter = TRUE, BoldHeaderRow = TRUE, na = "", FreezeRow = 1, FreezeCol = 1)
        
        save(Overexpressed,Underexpressed, file = file.path(RDataSupdir,paste0("Enrichment_test_",gp,"_vs_",ref,"_",annotbase,"_logFC",round(log2FC_threshold,2),".RData")))
    }
} 		
```

## Plot UMAP colored with differential genes and GSA enrichment

### Get gene lists and differential genes

```{r, echo=FALSE}
load(file=file.path(RDatadir,"LogCounts.RData"))
load(file=file.path(RDatadir,"annot_anocol_final.RData"))
load(file=file.path(RDataSupdir,"Supervised_res_object_edgeR.Rdata"))
annot_int_2 = annot_int[match(colnames(LogCounts),rownames(annot_int)),]

diff_genes <-  c("NNMT","TAGLN","INHBA","KRT6A","KRT6B","KRT16","KRT14","KLK10",
                 "KLK5","KLK7")
diff_genes_figures <- read.csv(file.path(maindir,"output/scRNAseq/PDX_BC976/CrossAnalysis_PDX","bivalent_persister_BC1224.csv"))  
diff_genes_figures = c("KLF4","FOXQ1","FOXN1","TFCP2L1")
unique(c(diff_genes,"CDKN2A","TGFBR3L","TGFBR1","KRT17", "INHBB", "KRT5","KRT8","BMP6","CDKN2B","TGFBR2",
                         "TGFBR3","TGFB2","INHBA","INHBB","SMAD2","TGFB1",
                         "FOSL1","NNMT","KRT14","KLK10","KLK5","TAGLN","NNMT",
                         "ELN","TGFBR3","MIF","ABCC4","ABCC3", my.res$Symbol[which(my.res$log2FC.persister > log2FC_threshold & 
                                                                                           my.res$qval.persister < Signif_threshold)]))


pcaText <- FALSE
annotText <- "sample_id"

gene_lists <- c("KEGG_TGF_BETA_SIGNALING_PATHWAY",
                "HALLMARK_EPITHELIAL_MESENCHYMAL_TRANSITION",
                "HALLMARK_APOPTOSIS","HALLMARK_TNFA","LEI_MYB_TARGETS")

annotCol = unique(c(annotCol,diff_genes,diff_genes_figures))
for(i in diff_genes_figures){
    if(i %in% rownames(LogCounts)) annot_int_2[,i] <- LogCounts[which(rownames(LogCounts)==i),]
}
gc()


anocol2 <- geco.unsupervised::geco.annotToCol4(
    annotS=annot_int_2,annotT=annot_int_2,
    plotLegend=T, plotLegendFile = file.path(resSUBdir,"Annotation_legends.pdf"),
    scale_q = "inferno")
for(i in intersect(colnames(anocol),colnames(anocol2))) anocol2[,i] = anocol[,i]

png(file.path(resdir_boxplots,"Boxplot_INHBA.png"),width=1500,height=1500,res=300)
boxplot(annot_int_2$INHBA~annot_int_2$sample_id,las=2)
dev.off()

save(anocol2,annot_int_2,file=file.path(RDataSupdir,"annot_anocol_final.RData")) 
```

### Generate figures: UMAP with correlation heatmap

```{r, out.width="30%", echo=FALSE}
load(file.path(RDatadir,"umap.RData"))
load(file.path(RDataSupdir,"annot_anocol_final.RData"))
load(file.path(RDatadir,"annot_anocol_final.RData"))

for(i in diff_genes_figures)
{
    j = which(colnames(anocol2)==i)
    png(file.path(resdir_UMAPs,paste0("UMAP_",colnames(anocol2)[j],".png")), height=1350,width=1200,res=300)
    if(class(annot_int_2[1,colnames(anocol2)[j]])=="numeric"){
        plot((umap_res), col=alpha(anocol2[,j],0.6),pch=20,cex=0.6,
             main=paste0(colnames(anocol2)[j]," min=",round(min(annot_int_2[,colnames(anocol2)[j]]),digits=3)," max=",round(max(annot_int_2[,colnames(anocol2)[j]]),digits=3)),
             xlab="component 1",ylab="component 2")
    } else {
        plot((umap_res), col=alpha(anocol2[,j],0.6),pch=20,cex=0.6,
             main=paste0(colnames(anocol2)[j]),
             xlab="component 1",ylab="component 2")
        if(colnames(anocol2)[j]=="sample_id"){
            #Plot persister with add plot since there is too low cell number
            pers = which(annot_int_2$sample_id=="BC1224_persister_6souris")
            points(umap_res[pers,], col=alpha(anocol2[pers,j],0.6),pch=20,cex=0.6)
            
        }
    }
    
    
    dev.off()
}
```

### Heatmap for a gene list of interest

```{r , out.width="30%",include=FALSE}
for(log2FC_threshold in log2FC_thresholds){
    tgfb_genes = unlist(str_split(Overexpressed$Deregulated_genes[which(Overexpressed$Gene.Set=="KEGG_TGF_BETA_SIGNALING_PATHWAY")],
                                  pattern=";"))
    mat <- LogCounts[rownames(LogCounts) %in% tgfb_genes,]
    tmat <- t(mat)
    
    distHC <- c("distPearson","distCosine","euclidean","maximum","manhattan","canberra","binary","minkowski")[3]
    methHC <- c("ward","ward.D","ward.D2","single","complete","average")[3]
    if(distHC=="distPearson")	d <- distPearson(tmat)
    if(distHC=="distCosine")	d <- distCosine(tmat)
    if(distHC %in% c("euclidean","maximum","manhattan","canberra","binary","minkowski"))	
        d <- dist(tmat,distHC)
    hc <- hclust(d,method=methHC)
    
    mat.so <- as.matrix(mat[,hc$order])
    
    if(chRangeHM){
        for(i in 1:nrow(mat.so)){
            mat.so[i,] <- geco.changeRange(mat.so[i,],newmin=0,newmax=1)
        }
    }
    
    png(file.path(resdir_heatmaps,paste0("Clustering_TGFB_logFC",round(log2FC_threshold,2),"_",distHC,"_",methHC,".png")), height=4000,width=4000,res=300)
    geco.hclustAnnotHeatmapPlot(x=(mat.so),
                                hc=hc,
                                hmColors=hmColors,
                                anocol=anocol2[hc$order,c(1:5)],#[,ncol(cc.col):1]
                                xpos=c(0.15,0.9,0.164,0.885),
                                ypos=c(0.1,0.5,0.5,0.6,0.62,0.95),
                                dendro.cex=0.01,
                                xlab.cex=1,
                                hmRowNames=TRUE,
                                hmRowNames.cex=0.5
    )
    dev.off()
}
```


###Study the top gene lists of persister cells

```{r , out.width="30%",include=FALSE}
for(log2FC_threshold in log2FC_thresholds){
    
    load(file.path(RDataSupdir,paste0("Enrichment_test_persister_6_vs_UNT_vs_UNT_MSigDB_logFC",round(log2FC_threshold,2),".RData")))
    
    how_many_top <- 20
    significPathway <- Overexpressed[Overexpressed$Class %in% c("c2_curated","c6_oncogenic","hallmark"),]
    
    significPathway <- significPathway[1:how_many_top,]
    significPathway_breast <- 	significPathway[grep(pattern = "BREAST", significPathway$Gene.Set),]
    
    
    #Calculate average expression for every cells on genes of each top pathways
    pathway_mat <- matrix(0,nrow=how_many_top,ncol=length(annot_int_2$sample_id))		
    pathway_names <- c()		
    for(i in 1:how_many_top){
        gene_list <- unlist(strsplit(as.character(significPathway$Deregulated_genes[i]),";"))
        mat <- LogCounts[gene_metadata$Symbol %in% gene_list,]
        pathway_mat[i,] <- apply(mat,2,mean) 
        pathway_names[i] <-significPathway$Gene.Set[i]
    }		
    
    row.names(pathway_mat) <- pathway_names
    tmat <- t(pathway_mat)
    
    selected_cells = annot_int_2$cell_id[which(annot_int_2$louvain_partition %in% c("C1","C5"))]
    rownames(tmat) = colnames(LogCounts)
    tmat = tmat[selected_cells,]
    distHC <- c("distPearson","distCosine","euclidean","maximum","manhattan","canberra","binary","minkowski")[3]
    methHC <- c("ward","ward.D","ward.D2","single","complete","average")[2]
    if(distHC=="distPearson")	d <- distPearson(tmat)
    if(distHC=="distCosine")	d <- distCosine(tmat)
    if(distHC %in% c("euclidean","maximum","manhattan","canberra","binary","minkowski"))	d <- dist(tmat,distHC)
    hc <- hclust(d,method=methHC)
    mat.so <- pathway_mat[,hc$order]
    
    if(chRangeHM){
        for(i in 1:nrow(mat.so)){
            mat.so[i,] <- geco.changeRange(mat.so[i,],newmin=0,newmax=1)
        }
    }
    rowClust <- hclust(as.dist(1 - cor(t(mat.so))), method = "ward.D")
    
    png(file.path(resdir_heatmaps,paste0("Clustering_Pathways_Persisters_homemade_logFC",log2FC_threshold,"_",distHC,"_",methHC,".png")), height=1000,width=1000,res=150)
    geco.hclustAnnotHeatmapPlot(x=(mat.so[rowClust$order,]),
                                hc=hc,
                                hmColors=hmColors,
                                anocol=anocol2[hc$order,c(1,8)],#[,ncol(cc.col):1]
                                xpos=c(0.15,0.9,0.164,0.885),
                                ypos=c(0.1,0.5,0.5,0.6,0.62,0.95),
                                dendro.cex=0.1,
                                xlab.cex=0.6,
                                hmRowNames=TRUE,
                                hmRowNames.cex=0.5
    )
    dev.off()
    nclust <- 5
    pdf(file.path(resdir_heatmaps,paste0("Clustering_Pathways_PersisterslogFC",log2FC_threshold,"_.pdf")))
    gplots::heatmap.2(pathway_mat,ColSideColors=anocol2[,1],trace="none")
    dev.off()
    
}
```
