---
title: "Pseudo-bulk peaks K27 HBCx95 UNT m43"
author: "CÃ©line Vallot"
date: "05/06/2019"
output:
  html_document: default
  pdf_document: default
---

```{r setup, include=FALSE}
# Create 10k transcript TSS K27 annotation 
library(here)
maindir = here()
source(file.path(maindir,"Scripts","global_var.R"))
data("hg38.GeneTSS")
```

```{r, echo=FALSE}
#Directories
dataDir <- file.path(maindir,"input","bulk_ChIPseq","Human")
annotDir <- file.path(maindir,"annotation")
resdir <- file.path(maindir,"output","bulk_ChIPseq","Human","Bivalence")
if(!dir.exists(resdir)) dir.create(resdir)

annot_10k = read.table(gzfile(file.path(maindir,"annotation",
                                        "gencode.v34.annotation.transcriptTSS_10k.bed.gz")),
                       sep="\t", header = F)
colnames(annot_10k) = c("chr","start","end","transcripts","gene","strand")
annot_10k = as(annot_10k,"GRanges")
annot_10k$ID = paste0(seqnames(annot_10k),"_",start(annot_10k),"_", end(annot_10k))

```


```{r, echo=FALSE}
# Count matrice
resultZ <- read.csv(file.path(maindir,"input","bulk_ChIPseq","Human","Count_Matrices","results_human_tumors_H3K4me3_H3K27me3_TSS.csv.gz"))
resultZ = resultZ[,grep("K27me3|K4me3",colnames(resultZ))]

loci = paste0(resultZ[,1],":",resultZ[,2],",",resultZ[,3])
RZ <- resultZ  #RZ matrix is the raw count matrix
RZ_RPKM <- (10^9*t(t((RZ)/10001/sum(RZ))))

samples = unique(gsub("|_K.*","",colnames(RZ_RPKM)))

over_PDX_BC976 = readxl::read_xlsx(file.path(maindir,"output","scRNAseq","PDX_BC976","Supervised","Tables","Differential_analysis_Limma_logFC_1.58.xlsx"))
over_PDX_BC408 = readxl::read_xlsx(file.path(maindir,"output","scRNAseq","PDX_BC408","Supervised","Tables","Differential_analysis_Limma_logFC_Pers_vs_chemonaive_1.58.xlsx"))
over_PDX_BC1224 = readxl::read_xlsx(file.path(maindir,"output","scRNAseq","PDX_BC1224","Supervised","Tables","Differential_analysis_Limma_logFC_Pers_vs_chemonaive_1.58.xlsx"))

intersection_with_scRNA = data.frame(sample=samples,n_bivalent_genes=0,n_BC976=0,n_BC408=0,n_BC1224=0, p.value_BC976= 0, p.value_BC408= 0,p.value_BC1224= 0)
for(i in samples){
  K4_samp = paste0(i,"_K4me3")
  K27_samp = paste0(i,"_K27me3")
  RZ_RPKM_K4 = RZ_RPKM[,K4_samp]
  RZ_RPKM_K27 = RZ_RPKM[,K27_samp]
  K27_K4_df = data.frame("Gene" = annot_10k$gene, "ID" = annot_10k$ID, "K27" = log2(RZ_RPKM_K27 + 1), "K4" = log2(RZ_RPKM_K4 + 1) )
  K27_K4_df = K27_K4_df[-grep("chrM",K27_K4_df$ID),]

  thresh_K4= quantile(K27_K4_df$K4,0.85)
  thresh_K27= quantile(K27_K4_df$K27,0.85)
# Corrplot + denisty
# Bin size control + color palette
png(file.path(resdir,paste0("Biplot_log2_K27_K4_10k_",i,".png")),height = 1200, width = 1200, res = 150)
hist_bottom <- ggplot(K27_K4_df) + geom_histogram(aes(K27), alpha = 0.65, bins = 150) + theme_classic()
hist_left <- ggplot(K27_K4_df)+geom_histogram(aes(K4), alpha = 0.65, bins = 150) + coord_flip() + theme_classic()
empty <- ggplot()+geom_point(aes(1,1), colour="white")+
         theme(axis.ticks=element_blank(),
               panel.background=element_blank(),
               axis.text.x=element_blank(), axis.text.y=element_blank(),
               axis.title.x=element_blank(), axis.title.y=element_blank())
top_biv_IDs <- c(head(K27_K4_df$ID[order(K27_K4_df$K27,decreasing=T)],n=7), head(K27_K4_df$ID[order(K27_K4_df$K4,decreasing=T)]))

scatter <- ggplot(K27_K4_df,aes(K27, K4)) + geom_point(alpha=0.65) + ggrepel::geom_label_repel(
    data = K27_K4_df %>% filter(ID %in% top_biv_IDs), # Filter data first
    aes(label=Gene)
  ) + theme_classic() +
  geom_hline(yintercept = thresh_K4, col = "red", lty = 2) +
  geom_vline(xintercept = thresh_K27, col = "red", lty = 2)
print(gridExtra::grid.arrange(hist_bottom, empty, scatter, hist_left, ncol=2, nrow=2, widths=c(4, 1), heights=c(1, 4)))
dev.off()

png(file.path(resdir,paste0("Biplot_log2_K27_K4_",i,".png")),height = 1200, width = 1200, res = 150)
print(smoothScatter(K27_K4_df$K27,K27_K4_df$K4,colramp=viridis,pch=19, xlab ="K27 - log2(RPKM)", ylab = "K4 - log2(log2(RPKM))"))
dev.off()

K27_K4_df_filt = K27_K4_df %>% filter(K27> thresh_K27 & K4> thresh_K4)

K27_K4_df_filt[grep("FOX",K27_K4_df_filt$Gene),]

bivalent_genes = unique(K27_K4_df_filt$Gene)
database <- MSIG.ls ##MSigDB
reflist <- union(unique(hg38.GeneTSS$Gene),bivalent_genes);length(reflist)

enrich.test <- geco.enrichmentTest(gene.sets=database,mylist=bivalent_genes,possibleIds=reflist)
enrich.test <- data.frame(Gene_set_name=rownames(enrich.test), enrich.test, check.names=FALSE)
enrich.test <- merge( subset(MSIG.gs, select=-Genes), enrich.test, by.x="Gene.Set", by.y="Gene_set_name", all.y=TRUE, sort=FALSE ) ## Get class of gene set
enrich.test <- enrich.test[order(enrich.test$`p-value`),]
enrich.test <- enrich.test[order(enrich.test$`p-value`),]
#ind <- which(enrich.test$`q-value`<= 0.1);if(!length(ind)){ind <- 1:20}
#Overexpressed  <- enrich.test[ind,]		}
Bivalent_pathways  <- enrich.test

WriteXLS(
    c("Bivalent_pathways"),
    ExcelFileName = file.path(resdir,
                              paste0("Enrichment_test_Bivalent_pathways_",i,".xlsx")),
    SheetNames = "Bivalent_pathways",
    perl = "perl", verbose = FALSE, row.names = FALSE, col.names = TRUE,
    AdjWidth = TRUE, AutoFilter = TRUE, BoldHeaderRow = TRUE, na = "",
    FreezeRow = 1, FreezeCol = 1)

# Fisher test
all_genes = hg38.GeneTSS$Gene
for(j in c("BC976","BC408","BC1224")){
  tmp = get(paste0("over_PDX_",j))
  over_genes = tmp$Symbol
  A_B = length(intersect(bivalent_genes,over_genes))
  A_nonB = length(bivalent_genes) - A_B
  B_nonA =  length(over_genes) - A_B
  Omega = length(all_genes) - length(union(bivalent_genes,over_genes))

  contingency_table_A_U_B = data.frame("Inside B" = c(A_B,
                                     B_nonA),
                               "Outside B" = c(A_nonB,
                                     Omega), row.names = c("A","All Genes"))

  intersection_with_scRNA[which(intersection_with_scRNA$sample==i),paste0("p.value_",j)] = fisher.test(contingency_table_A_U_B,alternative = "greater")$p.value
  intersection_with_scRNA[which(intersection_with_scRNA$sample==i),paste0("n_",j)] = A_B

}
intersection_with_scRNA$n_bivalent_genes[which(intersection_with_scRNA$sample==i)] = length(bivalent_genes)
K27_K4_df_filt$Significant = TRUE
K27_K4_df$Significant = FALSE
K27_K4_df$Significant[match(paste0(K27_K4_df_filt$Gene,K27_K4_df_filt$ID),
                            paste0(K27_K4_df$Gene,K27_K4_df$ID))] = TRUE
length(which(K27_K4_df_filt$Significant))
length(which(K27_K4_df$Significant))
K27_K4_df$Threshold_K27 = thresh_K27
K27_K4_df$Threshold_K4 = thresh_K4
write.csv(K27_K4_df %>% arrange(desc(Significant)), file.path(resdir,paste0("K27_K4_df_",i,".csv")))
}

write.csv(intersection_with_scRNA, file.path(resdir,paste0("Intersection_Bivalence_human_scRNA_PDX.csv")))

library(clusterProfiler)

```

# Intersection of all human bivalent pathways


```{r, echo=FALSE}
intersection_with_scRNA = read.csv(file.path(resdir,paste0("Intersection_Bivalence_human_scRNA_PDX.csv")))

list_gene_sets = list()
classes = c("c2_curated", "c5_GO","hallmark")

tab_init = readxl::read_xlsx(file.path(resdir, paste0("Enrichment_test_Bivalent_pathways_",samples[1],".xlsx")))
tab_init = tab_init[which(tab_init$Class %in% classes),]
# tab_init = tab_init[grep(pattern,tab_init$Class),]

tab_init = tab_init[order(tab_init$Gene.Set),]

for(i in samples){
  tab = readxl::read_xlsx(file.path(resdir, paste0("Enrichment_test_Bivalent_pathways_",i,".xlsx")))
  tab = tab[which(tab$Class %in% classes),]
  tab$`q-value` = as.numeric(tab$`q-value`)
  tab$log10.qvalue = -log10(tab$`q-value`)
  # tab = tab[which(tab$`q-value` < 0.1),]
  # tab = tab[grep(pattern, tab$Gene.Set),]
  tab$Cluster = i
  tab = tab[order(tab$Gene.Set),]
  tab = tab[,c(4,6,7,8)]
 list_gene_sets[[i]] = tab 
}

# combined_GS = do.call("rbind",list_gene_sets)
combined_GS = do.call("cbind",list_gene_sets)
combined_GS = cbind(tab_init[,1:3],combined_GS)
###
GSA_combined_save =  combined_GS
GSA_combined_save =  mutate(GSA_combined_save,
                            mean_qvalue = rowMeans(dplyr::select(GSA_combined_save, ends_with("q-value")),na.rm = T),
                            mean_log10.qvalue = rowMeans(dplyr::select(GSA_combined_save, ends_with("log10.qvalue")),na.rm = T),
                            mean_Nb_Deregulated_Genes = rowMeans(dplyr::select(GSA_combined_save, ends_with("Nb_of_deregulated_genes")),na.rm = T)
                            )

GSA_combined_save = GSA_combined_save[,c(1:3,40:42,4:39)]
GSA_combined_save = GSA_combined_save %>% arrange(mean_qvalue)

library(dplyr)
library(tidyr)

colnames(GSA_combined_save) = gsub("Deregulated_genes","Genes_Deregulated",colnames(GSA_combined_save))
GSA_combined_save = GSA_combined_save %>% dplyr::select(!ends_with(".log10.qvalue")) %>%
 pivot_longer(hT_PDX_BC1006.Nb_of_deregulated_genes:hT_PDX_BC976.Genes_Deregulated,
   names_to = c("Cluster",".value"),
   names_pattern = "(hT_PDX_BC.*)\\.(.*)"
 )

###
combined_GS = GSA_combined_save
combined_GS$n_bivalent_genes = intersection_with_scRNA$n_bivalent_genes[match(combined_GS$Cluster,
                                                                              intersection_with_scRNA$sample)]
combined_GS =  combined_GS %>% mutate(GeneRatio = paste0(Nb_of_deregulated_genes,"/", n_bivalent_genes))

combined_GS_filtered = combined_GS %>% filter(`q-value` < 0.1) %>% group_by(Gene.Set) %>%
  mutate(occurence =n(), mean_qval = mean(`q-value`)) %>% filter(occurence > 1)
combined_GS$N_sample_enriched = combined_GS_filtered$occurence[match(combined_GS$Gene.Set,combined_GS_filtered$Gene.Set)]

combined_GS = readxl::read_xlsx(file.path(resdir, ExcelFileName= "combined_GSA_bivalent_tumors_unfiltered_pivoted.xlsx"))

length(unique(combined_GS$Cluster))


# Breast or HALLMARK
pattern = "BREAST|MAMMARY|HALLMARK"
combined_GS_Breast_or_Hallmark = combined_GS[grep(pattern, combined_GS$Gene.Set),] %>% filter(N_sample_enriched > 6) %>% filter(`q-value` < 0.1)
length(unique(combined_GS_Breast_or_Hallmark$Gene.Set))

# combined_GS_C2_Mammary_Breast
pattern = "BREAST|MAMMARY"
combined_GS_C2_Mammary_Breast = combined_GS[grep(pattern, combined_GS$Gene.Set),] %>% filter(Class == "c2_curated" & N_sample_enriched > 6) %>% filter(`q-value` < 0.1)
length(unique(combined_GS_C2_Mammary_Breast$Gene.Set))

# combined_GS_C2_Mammary_Breast & HALLMARK # sorted by median qvalue
pattern = "BREAST|MAMMARY"
combined_GS_C2_Mammary_Breast_median = combined_GS[grep(pattern, combined_GS$Gene.Set),] %>% group_by(Gene.Set) %>% mutate(median_qvalue = median(`q-value`)) %>% ungroup %>% filter(Class == "c2_curated" & N_sample_enriched > 6) %>% filter(`q-value` < 0.1) %>% arrange(median_qvalue)

pattern = "HALLMARK"
combined_GS_HALLMARK_median = combined_GS[grep(pattern, combined_GS$Gene.Set),] %>% group_by(Gene.Set) %>% mutate(median_qvalue = median(`q-value`)) %>% ungroup %>% filter( N_sample_enriched > 1) %>% filter(`q-value` < 0.1) %>% arrange(median_qvalue)

combined_GS_top5_BC_or_HALLMARK_by_median = 
  rbind(
    combined_GS_C2_Mammary_Breast_median[which(combined_GS_C2_Mammary_Breast_median$Gene.Set %in% head(unique(combined_GS_C2_Mammary_Breast_median$Gene.Set),5)),],
combined_GS_HALLMARK_median[which(combined_GS_HALLMARK_median$Gene.Set %in% head(unique(combined_GS_HALLMARK_median$Gene.Set),5)),]
              )

length(unique(combined_GS_top5_BC_or_HALLMARK_by_median$Gene.Set))


# KEGG
combined_GS_KEGG = combined_GS[grep("KEGG", combined_GS$Gene.Set),] %>% filter(N_sample_enriched > 1)  %>% filter(`q-value` < 0.1)
length(unique(combined_GS_KEGG$Gene.Set))

# HALLMARK
combined_GS_HALLMARK = combined_GS[grep("HALLMARK", combined_GS$Gene.Set),] %>% filter(N_sample_enriched > 1)  %>% filter(`q-value` < 0.1)
length(unique(combined_GS_HALLMARK$Gene.Set))

# GO
combined_GS_GO = combined_GS[grep("c5_GO", combined_GS$Class),] %>% filter(N_sample_enriched > 6) %>% filter(`q-value` < 0.1)
length(unique(combined_GS_GO$Gene.Set))

# Top 50 among all below 500
combined_GS_TOP = combined_GS %>% filter( Nb_of_genes < 500, N_sample_enriched > 6) %>% filter(`q-value` < 0.1)
length(unique(combined_GS_TOP$Gene.Set))


# specific_order = c(
#   grep("HALLMARK",combined_GS_filtered$Gene.Set),
#   which(combined_GS_filtered$Gene.Set %in% grep("BREAST|MAMMARY",grep("HALLMARK|AMPLICON",combined_GS$Gene.Set, value = T, invert = T),value = T)),
#   grep("AMPLICON",combined_GS_filtered$Gene.Set)
# )
# combined_GS_filtered = combined_GS_filtered[specific_order,]

# Fake a cluster profiler object
library(clusterProfiler)
ck. <- new("compareClusterResult")


# Plot dotplot Breast or HALLMARK
png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_Breast_or_Hallmark.png"),width = 2000, height = 2500, res = 200)
ck = combined_GS_Breast_or_Hallmark[,c(7,1,1,12,12,9,9,9,10,4)]
colnames(ck) =c("Cluster","ID","Description","GeneRatio","BgRatio", "pvalue","p.adjust","qvalue","geneID", "Count") 
ck.@compareClusterResult = as.data.frame(ck)
clusterProfiler::dotplot(ck.,  showCategory = 5) + theme(axis.text.x = element_text(angle = 90)) + xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()

png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_Breast_or_Hallmark_HQ.png"),width = 2000, height = 3500, res = 400)
ck.@compareClusterResult = ck
clusterProfiler::dotplot(ck., showCategory = 5) + theme_classic() + theme(axis.text = element_blank()) + 
  xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()

# Plot dotplot C2 Mammary | breast 
png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_C2_Mammary_Breast.png"),width = 2000, height = 2500, res = 200)
ck = combined_GS_C2_Mammary_Breast[,c(7,1,1,12,12,9,9,9,10,4)]
colnames(ck) =c("Cluster","ID","Description","GeneRatio","BgRatio", "pvalue","p.adjust","qvalue","geneID", "Count") 
ck.@compareClusterResult = as.data.frame(ck)
clusterProfiler::dotplot(ck.,  showCategory = 5) + theme(axis.text.x = element_text(angle = 90)) + xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()

png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_C2_Mammary_Breast_HQ.png"),width = 2000, height = 3500, res = 400)
ck.@compareClusterResult = ck
clusterProfiler::dotplot(ck., showCategory = 5) + theme_classic() + theme(axis.text = element_blank()) + 
  xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()

# Plot dotplot C2 Mammary | breast arrange by median qvalue
png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_C2_Mammary_Breast_median.png"),width = 2000, height = 2500, res = 200)
ck = combined_GS_C2_Mammary_Breast_median[,c(7,1,1,12,12,9,9,9,10,4)]
colnames(ck) =c("Cluster","ID","Description","GeneRatio","BgRatio", "pvalue","p.adjust","qvalue","geneID", "Count") 
ck.@compareClusterResult = as.data.frame(ck)
clusterProfiler::dotplot(ck.,  showCategory = 5) + theme(axis.text.x = element_text(angle = 90)) + xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()

png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_C2_Mammary_Breast_median_HQ.png"),width = 2000, height = 3500, res = 400)
ck.@compareClusterResult = ck
clusterProfiler::dotplot(ck., showCategory = 5) + theme_classic() + theme(axis.text = element_blank()) + 
  xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()

# Plot dotplot top 5 by median C2 Mammary | breast & HALLMARK arranged by median qvalue
png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_C2_Mammary_Breast_or_HALLMARK_median.png"),width = 2000, height = 2500, res = 200)
ck = combined_GS_top5_BC_or_HALLMARK_by_median[,c(7,1,1,12,12,9,9,9,10,4)]
colnames(ck) =c("Cluster","ID","Description","GeneRatio","BgRatio", "pvalue","p.adjust","qvalue","geneID", "Count") 
ck.@compareClusterResult = as.data.frame(ck)
clusterProfiler::dotplot(ck.,  showCategory = 5) + theme(axis.text.x = element_text(angle = 90)) + xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()

png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_C2_Mammary_Breast_or_HALLMARK_median_HQ.png"),width = 2000, height = 3500, res = 400)
ck.@compareClusterResult = ck
clusterProfiler::dotplot(ck., showCategory = 5) + theme_classic() + theme(axis.text = element_blank()) + 
  xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()


# Plot dotplot KEGG
png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_KEGG.png"),width = 2000, height = 2500, res = 200)
ck = combined_GS_KEGG[,c(7,1,1,12,12,9,9,9,10,4)]
colnames(ck) =c("Cluster","ID","Description","GeneRatio","BgRatio", "pvalue","p.adjust","qvalue","geneID", "Count") 
ck.@compareClusterResult = as.data.frame(ck)
clusterProfiler::dotplot(ck.,  showCategory = 5) + theme(axis.text.x = element_text(angle = 90)) + xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()

png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_KEGG_HQ.png"),width = 2000, height = 3500, res = 400)
ck.@compareClusterResult = ck
clusterProfiler::dotplot(ck., showCategory = 5) + theme_classic() + theme(axis.text = element_blank()) + 
  xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()


# Plot dotplot GO
png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_GO.png"),width = 2000, height = 2500, res = 200)
ck = combined_GS_GO[,c(7,1,1,12,12,9,9,9,10,4)]
colnames(ck) =c("Cluster","ID","Description","GeneRatio","BgRatio", "pvalue","p.adjust","qvalue","geneID", "Count") 
ck.@compareClusterResult = as.data.frame(ck)
clusterProfiler::dotplot(ck.,  showCategory = 5) + theme(axis.text.x = element_text(angle = 90)) + xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()

png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_GO_HQ.png"),width = 2000, height = 3500, res = 400)
ck.@compareClusterResult = ck
clusterProfiler::dotplot(ck., showCategory = 5) + theme_classic() + theme(axis.text = element_blank()) + 
  xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()


# Plot dotplot HALLMARK
png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_HALLMARK.png"),width = 2000, height = 2500, res = 200)
ck = combined_GS_HALLMARK[,c(7,1,1,12,12,9,9,9,10,4)]
colnames(ck) =c("Cluster","ID","Description","GeneRatio","BgRatio", "pvalue","p.adjust","qvalue","geneID", "Count") 
ck.@compareClusterResult = as.data.frame(ck)
clusterProfiler::dotplot(ck.,  showCategory = 5) + theme(axis.text.x = element_text(angle = 90)) + xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()

png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_HALLMARK_HQ.png"),width = 2000, height = 3500, res = 400)
ck.@compareClusterResult = ck
clusterProfiler::dotplot(ck., showCategory = 5) + theme_classic() + theme(axis.text = element_blank()) + 
  xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()


# Plot dotplot top below 500
png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_top_all_below_500.png"),width = 2000, height = 2500, res = 200)
ck = combined_GS_TOP[,c(7,1,1,12,12,9,9,9,10,4)]
colnames(ck) =c("Cluster","ID","Description","GeneRatio","BgRatio", "pvalue","p.adjust","qvalue","geneID", "Count") 
ck.@compareClusterResult = as.data.frame(ck)
clusterProfiler::dotplot(ck.,  showCategory = 5) + theme(axis.text.x = element_text(angle = 90)) + xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()

png(filename = file.path(resdir,"DotPlot_GSA_human_tumors_top_all_below_500_HQ.png"),width = 2000, height = 3500, res = 400)
ck.@compareClusterResult = ck
clusterProfiler::dotplot(ck., showCategory = 5) + theme_classic() + theme(axis.text = element_blank()) + 
  xlab("") + scale_colour_gradientn(colours = (viridis::cividis(256)))
dev.off()


WriteXLS::WriteXLS(GSA_combined_save, ExcelFileName = file.path(resdir,"combined_GSA_bivalent_tumors_unfiltered.xlsx"))
WriteXLS::WriteXLS(combined_GS, col.names = T, row.names = F,)

save(ck, file = file.path(resdir,"combined_GSA_bivalent_tumors_ck.RData"))
```



